// motion-dom/dist/es/value/types/complex/index.js
import { sanitize as e } from "../utils/sanitize.js";
import { floatRegex as t } from "../utils/float-regex.js";
import { color as n } from "../color/index.js";
import { colorRegex as r } from "../utils/color-regex.js";
//#region node_modules/motion-dom/dist/es/value/types/complex/index.mjs
function i(e) {
	return isNaN(e) && typeof e == "string" && (e.match(t)?.length || 0) + (e.match(r)?.length || 0) > 0;
}
var a = "number", o = "color", s = "var", c = "var(", l = "${}", u = /var\s*\(\s*--(?:[\w-]+\s*|[\w-]+\s*,(?:\s*[^)(\s]|\s*\((?:[^)(]|\([^)(]*\))*\))+\s*)\)|#[\da-f]{3,8}|(?:rgb|hsl)a?\((?:-?[\d.]+%?[,\s]+){2}-?[\d.]+%?\s*(?:[,/]\s*)?(?:\b\d+(?:\.\d+)?|\.\d+)?%?\)|-?(?:\d+(?:\.\d+)?|\.\d+)/giu;
function d(e) {
	let t = e.toString(), r = [], i = {
		color: [],
		number: [],
		var: []
	}, d = [], f = 0;
	return {
		values: r,
		split: t.replace(u, (e) => (n.test(e) ? (i.color.push(f), d.push(o), r.push(n.parse(e))) : e.startsWith(c) ? (i.var.push(f), d.push(s), r.push(e)) : (i.number.push(f), d.push(a), r.push(parseFloat(e))), ++f, l)).split(l),
		indexes: i,
		types: d
	};
}
function f(e) {
	return d(e).values;
}
function p({ split: t, types: r }) {
	let i = t.length;
	return (s) => {
		let c = "";
		for (let l = 0; l < i; l++) if (c += t[l], s[l] !== void 0) {
			let t = r[l];
			t === a ? c += e(s[l]) : t === o ? c += n.transform(s[l]) : c += s[l];
		}
		return c;
	};
}
function m(e) {
	return p(d(e));
}
var h = (e) => typeof e == "number" ? 0 : n.test(e) ? n.getAnimatableNone(e) : e, g = (e, t) => typeof e == "number" ? t?.trim().endsWith("/") ? e : 0 : h(e);
function _(e) {
	let t = d(e);
	return p(t)(t.values.map((e, n) => g(e, t.split[n])));
}
var v = {
	test: i,
	parse: f,
	createTransformer: m,
	getAnimatableNone: _
};
//#endregion
export { d as analyseComplexValue, v as complex };
