// motion-dom/dist/es/projection/geometry/delta-apply.js
import { mixNumber as e } from "../../utils/mix/number.js";
import { hasTransform as t } from "../utils/has-transform.js";
//#region node_modules/motion-dom/dist/es/projection/geometry/delta-apply.mjs
function n(e, t, n) {
	return n + t * (e - n);
}
function r(e, t, r, i, a) {
	return a !== void 0 && (e = n(e, a, i)), n(e, r, i) + t;
}
function i(e, t = 0, n = 1, i, a) {
	e.min = r(e.min, t, n, i, a), e.max = r(e.max, t, n, i, a);
}
function a(e, { x: t, y: n }) {
	i(e.x, t.translate, t.scale, t.originPoint), i(e.y, n.translate, n.scale, n.originPoint);
}
var o = .999999999999, s = 1.0000000000001;
function c(e, n, r, i = !1) {
	let c = r.length;
	if (!c) return;
	n.x = n.y = 1;
	let u, d;
	for (let o = 0; o < c; o++) {
		u = r[o], d = u.projectionDelta;
		let { visualElement: s } = u.options;
		s && s.props.style && s.props.style.display === "contents" || (i && u.options.layoutScroll && u.scroll && u !== u.root && (l(e.x, -u.scroll.offset.x), l(e.y, -u.scroll.offset.y)), d && (n.x *= d.x.scale, n.y *= d.y.scale, a(e, d)), i && t(u.latestValues) && f(e, u.latestValues, u.layout?.layoutBox));
	}
	n.x < s && n.x > o && (n.x = 1), n.y < s && n.y > o && (n.y = 1);
}
function l(e, t) {
	e.min += t, e.max += t;
}
function u(t, n, r, a, o = .5) {
	i(t, n, r, e(t.min, t.max, o), a);
}
function d(e, t) {
	return typeof e == "string" ? parseFloat(e) / 100 * (t.max - t.min) : e;
}
function f(e, t, n) {
	let r = n ?? e;
	u(e.x, d(t.x, r.x), t.scaleX, t.scale, t.originX), u(e.y, d(t.y, r.y), t.scaleY, t.scale, t.originY);
}
//#endregion
export { a as applyBoxDelta, c as applyTreeDeltas, n as scalePoint, f as transformBox, l as translateAxis };
