// motion-dom/dist/es/projection/geometry/delta-calc.js
import { mixNumber as e } from "../../utils/mix/number.js";
//#region node_modules/motion-dom/dist/es/projection/geometry/delta-calc.mjs
var t = .9999, n = 1.0001, r = -.01, i = .01;
function a(e) {
	return e.max - e.min;
}
function o(e, t, n) {
	return Math.abs(e - t) <= n;
}
function s(o, s, c, l = .5) {
	o.origin = l, o.originPoint = e(s.min, s.max, o.origin), o.scale = a(c) / a(s), o.translate = e(c.min, c.max, o.origin) - o.originPoint, (o.scale >= t && o.scale <= n || isNaN(o.scale)) && (o.scale = 1), (o.translate >= r && o.translate <= i || isNaN(o.translate)) && (o.translate = 0);
}
function c(e, t, n, r) {
	s(e.x, t.x, n.x, r ? r.originX : void 0), s(e.y, t.y, n.y, r ? r.originY : void 0);
}
function l(t, n, r, i = 0) {
	t.min = (i ? e(r.min, r.max, i) : r.min) + n.min, t.max = t.min + a(n);
}
function u(e, t, n, r) {
	l(e.x, t.x, n.x, r?.x), l(e.y, t.y, n.y, r?.y);
}
function d(t, n, r, i = 0) {
	let o = i ? e(r.min, r.max, i) : r.min;
	t.min = n.min - o, t.max = t.min + a(n);
}
function f(e, t, n, r) {
	d(e.x, t.x, n.x, r?.x), d(e.y, t.y, n.y, r?.y);
}
//#endregion
export { c as calcBoxDelta, a as calcLength, u as calcRelativeBox, f as calcRelativePosition, o as isNear };
