// motion-dom/dist/es/render/html/utils/build-transform.js
import { transformPropOrder as e } from "../../utils/keys-transform.js";
import { numberValueTypes as t } from "../../../value/types/maps/number.js";
import { getValueAsType as n } from "../../../value/types/utils/get-as-type.js";
//#region node_modules/motion-dom/dist/es/render/html/utils/build-transform.mjs
var r = {
	x: "translateX",
	y: "translateY",
	z: "translateZ",
	transformPerspective: "perspective"
}, i = e.length;
function a(a, o, s) {
	let c = "", l = !0;
	for (let u = 0; u < i; u++) {
		let i = e[u], d = a[i];
		if (d === void 0) continue;
		let f = !0;
		if (typeof d == "number") f = d === +!!i.startsWith("scale");
		else {
			let e = parseFloat(d);
			f = i.startsWith("scale") ? e === 1 : e === 0;
		}
		if (!f || s) {
			let e = n(d, t[i]);
			if (!f) {
				l = !1;
				let t = r[i] || i;
				c += `${t}(${e}) `;
			}
			s && (o[i] = e);
		}
	}
	let u = a.pathRotation;
	return u && (l = !1, c += `rotate(${n(u, t.pathRotation)}) `), c = c.trim(), s ? c = s(o, l ? "" : c) : l && (c = "none"), c;
}
//#endregion
export { a as buildTransform };
