// motion-dom/dist/es/animation/generators/inertia.js
import { spring as e } from "./spring.js";
import { getGeneratorVelocity as t } from "./utils/velocity.js";
//#region node_modules/motion-dom/dist/es/animation/generators/inertia.mjs
function n({ keyframes: n, velocity: r = 0, power: i = .8, timeConstant: a = 325, bounceDamping: o = 10, bounceStiffness: s = 500, modifyTarget: c, min: l, max: u, restDelta: d = .5, restSpeed: f }) {
	let p = n[0], m = {
		done: !1,
		value: p
	}, h = (e) => l !== void 0 && e < l || u !== void 0 && e > u, g = (e) => l === void 0 ? u : u === void 0 || Math.abs(l - e) < Math.abs(u - e) ? l : u, _ = i * r, v = p + _, y = c === void 0 ? v : c(v);
	y !== v && (_ = y - p);
	let b = (e) => -_ * Math.exp(-e / a), x = (e) => y + b(e), S = (e) => {
		let t = b(e), n = x(e);
		m.done = Math.abs(t) <= d, m.value = m.done ? y : n;
	}, C, w, T = (n) => {
		h(m.value) && (C = n, w = e({
			keyframes: [m.value, g(m.value)],
			velocity: t(x, n, m.value),
			damping: o,
			stiffness: s,
			restDelta: d,
			restSpeed: f
		}));
	};
	return T(0), {
		calculatedDuration: null,
		next: (e) => {
			let t = !1;
			return !w && C === void 0 && (t = !0, S(e), T(e)), C !== void 0 && e >= C ? w.next(e - C) : (!t && S(e), m);
		}
	};
}
//#endregion
export { n as inertia };
