// motion-dom/dist/es/animation/generators/spring.js
import { clamp as e } from "../../../../../motion-utils/dist/es/clamp.js";
import { warning as t } from "../../../../../motion-utils/dist/es/errors.js";
import { millisecondsToSeconds as n, secondsToMilliseconds as r } from "../../../../../motion-utils/dist/es/time-conversion.js";
import { generateLinearEasing as i } from "../waapi/utils/linear.js";
import { calcGeneratorDuration as a, maxGeneratorDuration as o } from "./utils/calc-duration.js";
import { createGeneratorEasing as s } from "./utils/create-generator-easing.js";
//#region node_modules/motion-dom/dist/es/animation/generators/spring.mjs
var c = {
	stiffness: 100,
	damping: 10,
	mass: 1,
	velocity: 0,
	duration: 800,
	bounce: .3,
	visualDuration: .3,
	restSpeed: {
		granular: .01,
		default: 2
	},
	restDelta: {
		granular: .005,
		default: .5
	},
	minDuration: .01,
	maxDuration: 10,
	minDamping: .05,
	maxDamping: 1
};
function l(e, t) {
	return e * Math.sqrt(1 - t * t);
}
var u = 12;
function d(e, t, n) {
	let r = n;
	for (let n = 1; n < u; n++) r -= e(r) / t(r);
	return r;
}
var f = .001;
function p({ duration: i = c.duration, bounce: a = c.bounce, velocity: o = c.velocity, mass: s = c.mass }) {
	let u, p;
	t(i <= r(c.maxDuration), "Spring duration must be 10 seconds or less", "spring-duration-limit");
	let m = 1 - a;
	m = e(c.minDamping, c.maxDamping, m), i = e(c.minDuration, c.maxDuration, n(i)), m < 1 ? (u = (e) => {
		let t = e * m, n = t * i, r = t - o, a = l(e, m), s = Math.exp(-n);
		return f - r / a * s;
	}, p = (e) => {
		let t = e * m * i, n = t * o + o, r = m ** 2 * e ** 2 * i, a = Math.exp(-t), s = l(e ** 2, m);
		return (-u(e) + f > 0 ? -1 : 1) * ((n - r) * a) / s;
	}) : (u = (e) => -.001 + Math.exp(-e * i) * ((e - o) * i + 1), p = (e) => Math.exp(-e * i) * ((o - e) * (i * i)));
	let h = 5 / i, g = d(u, p, h);
	if (i = r(i), isNaN(g)) return {
		stiffness: c.stiffness,
		damping: c.damping,
		duration: i
	};
	{
		let e = g ** 2 * s;
		return {
			stiffness: e,
			damping: m * 2 * Math.sqrt(s * e),
			duration: i
		};
	}
}
var m = ["duration", "bounce"], h = [
	"stiffness",
	"damping",
	"mass"
];
function g(e, t) {
	return t.some((t) => e[t] !== void 0);
}
function _(t) {
	let n = {
		velocity: c.velocity,
		stiffness: c.stiffness,
		damping: c.damping,
		mass: c.mass,
		isResolvedFromDuration: !1,
		...t
	};
	if (!g(t, h) && g(t, m)) if (n.velocity = 0, t.visualDuration) {
		let r = t.visualDuration, i = 2 * Math.PI / (r * 1.2), a = i * i, o = 2 * e(.05, 1, 1 - (t.bounce || 0)) * Math.sqrt(a);
		n = {
			...n,
			mass: c.mass,
			stiffness: a,
			damping: o
		};
	} else {
		let e = p({
			...t,
			velocity: 0
		});
		n = {
			...n,
			...e,
			mass: c.mass
		}, n.isResolvedFromDuration = !0;
	}
	return n;
}
function v(e = c.visualDuration, t = c.bounce) {
	let s = typeof e == "object" ? e : {
		visualDuration: e,
		keyframes: [0, 1],
		bounce: t
	}, { restSpeed: u, restDelta: d } = s, f = s.keyframes[0], p = s.keyframes[s.keyframes.length - 1], m = {
		done: !1,
		value: f
	}, { stiffness: h, damping: g, mass: v, duration: y, velocity: b, isResolvedFromDuration: x } = _({
		...s,
		velocity: -n(s.velocity || 0)
	}), S = b || 0, C = g / (2 * Math.sqrt(h * v)), w = p - f, T = n(Math.sqrt(h / v)), E = Math.abs(w) < 5;
	u ||= E ? c.restSpeed.granular : c.restSpeed.default, d ||= E ? c.restDelta.granular : c.restDelta.default;
	let D, O, k, A, j, M;
	if (C < 1) k = l(T, C), A = (S + C * T * w) / k, D = (e) => p - Math.exp(-C * T * e) * (A * Math.sin(k * e) + w * Math.cos(k * e)), j = C * T * A + w * k, M = C * T * w - A * k, O = (e) => Math.exp(-C * T * e) * (j * Math.sin(k * e) + M * Math.cos(k * e));
	else if (C === 1) {
		D = (e) => p - Math.exp(-T * e) * (w + (S + T * w) * e);
		let e = S + T * w;
		O = (t) => Math.exp(-T * t) * (T * e * t - S);
	} else {
		let e = T * Math.sqrt(C * C - 1);
		D = (t) => {
			let n = Math.exp(-C * T * t), r = Math.min(e * t, 300);
			return p - n * ((S + C * T * w) * Math.sinh(r) + e * w * Math.cosh(r)) / e;
		};
		let t = (S + C * T * w) / e, n = C * T * t - w * e, r = C * T * w - t * e;
		O = (t) => {
			let i = Math.exp(-C * T * t), a = Math.min(e * t, 300);
			return i * (n * Math.sinh(a) + r * Math.cosh(a));
		};
	}
	let N = {
		calculatedDuration: x && y || null,
		velocity: (e) => r(O(e)),
		next: (e) => {
			if (!x && C < 1) {
				let t = Math.exp(-C * T * e), n = Math.sin(k * e), i = Math.cos(k * e), a = p - t * (A * n + w * i), o = r(t * (j * n + M * i));
				return m.done = Math.abs(o) <= u && Math.abs(p - a) <= d, m.value = m.done ? p : a, m;
			}
			let t = D(e);
			if (x) m.done = e >= y;
			else {
				let n = r(O(e));
				m.done = Math.abs(n) <= u && Math.abs(p - t) <= d;
			}
			return m.value = m.done ? p : t, m;
		},
		toString: () => {
			let e = Math.min(a(N), o), t = i((t) => N.next(e * t).value, e, 30);
			return e + "ms " + t;
		},
		toTransition: () => {}
	};
	return N;
}
v.applyToOptions = (e) => {
	let t = s(e, 100, v);
	return e.ease = t.ease, e.duration = r(t.duration), e.type = "keyframes", e;
};
//#endregion
export { v as spring };
