// motion-dom/dist/es/animation/AsyncMotionValueAnimation.js
import { MotionGlobalConfig as e } from "../../../../motion-utils/dist/es/global-config.js";
import { noop as t } from "../../../../motion-utils/dist/es/noop.js";
import { time as n } from "../frameloop/sync-time.js";
import { getFinalKeyframe as r } from "./keyframes/get-final.js";
import { WithPromise as i } from "./utils/WithPromise.js";
import { JSAnimation as a } from "./JSAnimation.js";
import { KeyframeResolver as o, flushKeyframeResolvers as s } from "./keyframes/KeyframesResolver.js";
import { NativeAnimationExtended as c } from "./NativeAnimationExtended.js";
import { canAnimate as l } from "./utils/can-animate.js";
import { makeAnimationInstant as u } from "./utils/make-animation-instant.js";
import { supportsBrowserAnimation as d } from "./waapi/supports/waapi.js";
//#region node_modules/motion-dom/dist/es/animation/AsyncMotionValueAnimation.mjs
var f = 40, p = class extends i {
	constructor({ autoplay: e = !0, delay: t = 0, type: r = "keyframes", repeat: i = 0, repeatDelay: a = 0, repeatType: s = "loop", keyframes: c, name: l, motionValue: u, element: d, ...f }) {
		super(), this.stop = () => {
			this._animation && (this._animation.stop(), this.stopTimeline?.()), this.keyframeResolver?.cancel();
		}, this.createdAt = n.now();
		let p = {
			autoplay: e,
			delay: t,
			type: r,
			repeat: i,
			repeatDelay: a,
			repeatType: s,
			name: l,
			motionValue: u,
			element: d,
			...f
		}, m = d?.KeyframeResolver || o;
		this.keyframeResolver = new m(c, (e, t, n) => this.onKeyframesResolved(e, t, p, !n), l, u, d), this.keyframeResolver?.scheduleResolve();
	}
	onKeyframesResolved(i, o, s, p) {
		this.keyframeResolver = void 0;
		let { name: m, type: h, velocity: g, delay: _, isHandoff: v, onUpdate: y } = s;
		this.resolvedAt = n.now();
		let b = !0;
		l(i, m, h, g) || (b = !1, (e.instantAnimations || !_) && y?.(r(i, s, o)), i[0] = i[i.length - 1], u(s), s.repeat = 0);
		let x = {
			startTime: p ? this.resolvedAt && this.resolvedAt - this.createdAt > f ? this.resolvedAt : this.createdAt : void 0,
			finalKeyframe: o,
			...s,
			keyframes: i
		}, S = b && !v && d(x), C = x.motionValue?.owner?.current, w;
		if (S) try {
			w = new c({
				...x,
				element: C
			});
		} catch {
			w = new a(x);
		}
		else w = new a(x);
		w.finished.then(() => {
			this.notifyFinished();
		}).catch(t), this.pendingTimeline &&= (this.stopTimeline = w.attachTimeline(this.pendingTimeline), void 0), this._animation = w;
	}
	get finished() {
		return this._animation ? this.animation.finished : this._finished;
	}
	then(e, t) {
		return this.finished.finally(e).then(() => {});
	}
	get animation() {
		return this._animation || (this.keyframeResolver?.resume(), s()), this._animation;
	}
	get duration() {
		return this.animation.duration;
	}
	get iterationDuration() {
		return this.animation.iterationDuration;
	}
	get time() {
		return this.animation.time;
	}
	set time(e) {
		this.animation.time = e;
	}
	get speed() {
		return this.animation.speed;
	}
	get state() {
		return this.animation.state;
	}
	set speed(e) {
		this.animation.speed = e;
	}
	get startTime() {
		return this.animation.startTime;
	}
	attachTimeline(e) {
		return this._animation ? this.stopTimeline = this.animation.attachTimeline(e) : this.pendingTimeline = e, () => this.stop();
	}
	play() {
		this.animation.play();
	}
	pause() {
		this.animation.pause();
	}
	complete() {
		this.animation.complete();
	}
	cancel() {
		this._animation && this.animation.cancel(), this.keyframeResolver?.cancel();
	}
};
//#endregion
export { p as AsyncMotionValueAnimation };
