// motion-dom/dist/es/animation/JSAnimation.js
import { clamp as e } from "../../../../motion-utils/dist/es/clamp.js";
import { invariant as t } from "../../../../motion-utils/dist/es/errors.js";
import { pipe as n } from "../../../../motion-utils/dist/es/pipe.js";
import { millisecondsToSeconds as r, secondsToMilliseconds as i } from "../../../../motion-utils/dist/es/time-conversion.js";
import { time as a } from "../frameloop/sync-time.js";
import { activeAnimations as o } from "../stats/animation-count.js";
import { mix as s } from "../utils/mix/index.js";
import { frameloopDriver as c } from "./drivers/frame.js";
import { calcGeneratorDuration as l } from "./generators/utils/calc-duration.js";
import { getGeneratorVelocity as u } from "./generators/utils/velocity.js";
import { inertia as d } from "./generators/inertia.js";
import { keyframes as f } from "./generators/keyframes.js";
import { getFinalKeyframe as p } from "./keyframes/get-final.js";
import { replaceTransitionType as m } from "./utils/replace-transition-type.js";
import { WithPromise as h } from "./utils/WithPromise.js";
//#region node_modules/motion-dom/dist/es/animation/JSAnimation.mjs
var g = (e) => e / 100, _ = class extends h {
	constructor(e) {
		super(), this.state = "idle", this.startTime = null, this.isStopped = !1, this.currentTime = 0, this.holdTime = null, this.playbackSpeed = 1, this.delayState = {
			done: !1,
			value: void 0
		}, this.stop = () => {
			let { motionValue: e } = this.options;
			e && e.updatedAt !== a.now() && this.tick(a.now()), this.isStopped = !0, this.state !== "idle" && (this.teardown(), this.options.onStop?.());
		}, o.mainThread++, this.options = e, this.initAnimation(), this.play(), e.autoplay === !1 && this.pause();
	}
	initAnimation() {
		let { options: e } = this;
		m(e);
		let { type: r = f, repeat: i = 0, repeatDelay: a = 0, repeatType: o, velocity: c = 0 } = e, { keyframes: u } = e, d = r || f;
		process.env.NODE_ENV !== "production" && d !== f && t(u.length <= 2, `Only two keyframes currently supported with spring and inertia animations. Trying to animate ${u}`, "spring-two-frames"), d !== f && typeof u[0] != "number" && (this.mixKeyframes = n(g, s(u[0], u[1])), u = [0, 100]);
		let p = d({
			...e,
			keyframes: u
		});
		o === "mirror" && (this.mirroredGenerator = d({
			...e,
			keyframes: [...u].reverse(),
			velocity: -c
		})), p.calculatedDuration === null && (p.calculatedDuration = l(p));
		let { calculatedDuration: h } = p;
		this.calculatedDuration = h, this.resolvedDuration = h + a, this.totalDuration = this.resolvedDuration * (i + 1) - a, this.generator = p;
	}
	updateTime(e) {
		let t = Math.round(e - this.startTime) * this.playbackSpeed;
		this.holdTime === null ? this.currentTime = t : this.currentTime = this.holdTime;
	}
	tick(t, n = !1) {
		let { generator: r, totalDuration: i, mixKeyframes: a, mirroredGenerator: o, resolvedDuration: s, calculatedDuration: c } = this;
		if (this.startTime === null) return r.next(0);
		let { delay: l = 0, keyframes: u, repeat: f, repeatType: m, repeatDelay: h, type: g, onUpdate: _, finalKeyframe: v } = this.options;
		this.speed > 0 ? this.startTime = Math.min(this.startTime, t) : this.speed < 0 && (this.startTime = Math.min(t - i / this.speed, this.startTime)), n ? this.currentTime = t : this.updateTime(t);
		let y = this.currentTime - l * (this.playbackSpeed >= 0 ? 1 : -1), b = this.playbackSpeed >= 0 ? y < 0 : y > i;
		this.currentTime = Math.max(y, 0), this.state === "finished" && this.holdTime === null && (this.currentTime = i);
		let x = this.currentTime, S = r;
		if (f) {
			let t = Math.min(this.currentTime, i) / s, n = Math.floor(t), r = t % 1;
			!r && t >= 1 && (r = 1), r === 1 && n--, n = Math.min(n, f + 1), n % 2 && (m === "reverse" ? (r = 1 - r, h && (r -= h / s)) : m === "mirror" && (S = o)), x = e(0, 1, r) * s;
		}
		let C;
		b ? (this.delayState.value = u[0], C = this.delayState) : C = S.next(x), a && !b && (C.value = a(C.value));
		let { done: w } = C;
		!b && c !== null && (w = this.playbackSpeed >= 0 ? this.currentTime >= i : this.currentTime <= 0);
		let T = this.holdTime === null && (this.state === "finished" || this.state === "running" && w);
		return T && g !== d && (C.value = p(u, this.options, v, this.speed)), _ && _(C.value), T && this.finish(), C;
	}
	then(e, t) {
		return this.finished.then(e, t);
	}
	get duration() {
		return r(this.calculatedDuration);
	}
	get iterationDuration() {
		let { delay: e = 0 } = this.options || {};
		return this.duration + r(e);
	}
	get time() {
		return r(this.currentTime);
	}
	set time(e) {
		e = i(e), this.currentTime = e, this.startTime === null || this.holdTime !== null || this.playbackSpeed === 0 ? this.holdTime = e : this.driver && (this.startTime = this.driver.now() - e / this.playbackSpeed), this.driver ? this.driver.start(!1) : (this.startTime = 0, this.state = "paused", this.holdTime = e, this.tick(e));
	}
	getGeneratorVelocity() {
		let e = this.currentTime;
		if (e <= 0) return this.options.velocity || 0;
		if (this.generator.velocity) return this.generator.velocity(e);
		let t = this.generator.next(e).value;
		return u((e) => this.generator.next(e).value, e, t);
	}
	get speed() {
		return this.playbackSpeed;
	}
	set speed(e) {
		let t = this.playbackSpeed !== e;
		t && this.driver && this.updateTime(a.now()), this.playbackSpeed = e, t && this.driver && (this.time = r(this.currentTime));
	}
	play() {
		if (this.isStopped) return;
		let { driver: e = c, startTime: t } = this.options;
		this.driver ||= e((e) => this.tick(e)), this.options.onPlay?.();
		let n = this.driver.now();
		this.state === "finished" ? (this.updateFinished(), this.startTime = n) : this.holdTime === null ? this.startTime ||= t ?? n : this.startTime = n - this.holdTime, this.state === "finished" && this.speed < 0 && (this.startTime += this.calculatedDuration), this.holdTime = null, this.state = "running", this.driver.start();
	}
	pause() {
		this.state = "paused", this.updateTime(a.now()), this.holdTime = this.currentTime;
	}
	complete() {
		this.state !== "running" && this.play(), this.state = "finished", this.holdTime = null;
	}
	finish() {
		this.notifyFinished(), this.teardown(), this.state = "finished", this.options.onComplete?.();
	}
	cancel() {
		this.holdTime = null, this.startTime = 0, this.tick(0), this.teardown(), this.options.onCancel?.();
	}
	teardown() {
		this.state = "idle", this.stopDriver(), this.startTime = this.holdTime = null, o.mainThread--;
	}
	stopDriver() {
		this.driver &&= (this.driver.stop(), void 0);
	}
	sample(e) {
		return this.startTime = 0, this.tick(e, !0);
	}
	attachTimeline(e) {
		return this.options.allowFlatten && (this.options.type = "keyframes", this.options.ease = "linear", this.initAnimation()), this.driver?.stop(), e.observe(this);
	}
};
//#endregion
export { _ as JSAnimation };
