// motion-dom/dist/es/animation/NativeAnimation.js
import { invariant as e } from "../../../../motion-utils/dist/es/errors.js";
import { noop as t } from "../../../../motion-utils/dist/es/noop.js";
import { millisecondsToSeconds as n, secondsToMilliseconds as r } from "../../../../motion-utils/dist/es/time-conversion.js";
import { getFinalKeyframe as i } from "./keyframes/get-final.js";
import { WithPromise as a } from "./utils/WithPromise.js";
import { setStyle as o } from "../render/dom/style-set.js";
import { supportsScrollTimeline as s } from "../utils/supports/scroll-timeline.js";
import { startWaapiAnimation as c } from "./waapi/start-waapi-animation.js";
import { applyGeneratorOptions as l } from "./waapi/utils/apply-generator.js";
//#region node_modules/motion-dom/dist/es/animation/NativeAnimation.mjs
var u = class extends a {
	constructor(t) {
		if (super(), this.finishedTime = null, this.isStopped = !1, this.manualStartTime = null, !t) return;
		let { element: n, name: r, keyframes: a, pseudoElement: s, allowFlatten: u = !1, finalKeyframe: d, onComplete: f } = t;
		this.isPseudoElement = !!s, this.allowFlatten = u, this.options = t, e(typeof t.type != "string", "Mini animate() doesn't support \"type\" as a string.", "mini-spring");
		let p = l(t);
		this.animation = c(n, r, a, p, s), p.autoplay === !1 && this.animation.pause(), this.animation.onfinish = () => {
			if (this.finishedTime = this.time, !s) {
				let e = i(a, this.options, d, this.speed);
				this.updateMotionValue && this.updateMotionValue(e), o(n, r, e), this.animation.cancel();
			}
			f?.(), this.notifyFinished();
		};
	}
	play() {
		this.isStopped || (this.manualStartTime = null, this.animation.play(), this.state === "finished" && this.updateFinished());
	}
	pause() {
		this.animation.pause();
	}
	complete() {
		this.animation.finish?.();
	}
	cancel() {
		try {
			this.animation.cancel();
		} catch {}
	}
	stop() {
		if (this.isStopped) return;
		this.isStopped = !0;
		let { state: e } = this;
		e === "idle" || e === "finished" || (this.updateMotionValue ? this.updateMotionValue() : this.commitStyles(), this.isPseudoElement || this.cancel());
	}
	commitStyles() {
		let e = this.options?.element;
		!this.isPseudoElement && e?.isConnected && this.animation.commitStyles?.();
	}
	get duration() {
		let e = this.animation.effect?.getComputedTiming?.().duration || 0;
		return n(Number(e));
	}
	get iterationDuration() {
		let { delay: e = 0 } = this.options || {};
		return this.duration + n(e);
	}
	get time() {
		return n(Number(this.animation.currentTime) || 0);
	}
	set time(e) {
		let t = this.finishedTime !== null;
		this.manualStartTime = null, this.finishedTime = null, this.animation.currentTime = r(e), t && this.animation.pause();
	}
	get speed() {
		return this.animation.playbackRate;
	}
	set speed(e) {
		e < 0 && (this.finishedTime = null), this.animation.playbackRate = e;
	}
	get state() {
		return this.finishedTime === null ? this.animation.playState : "finished";
	}
	get startTime() {
		return this.manualStartTime ?? Number(this.animation.startTime);
	}
	set startTime(e) {
		this.manualStartTime = this.animation.startTime = e;
	}
	attachTimeline({ timeline: e, rangeStart: n, rangeEnd: r, observe: i }) {
		return this.allowFlatten && this.animation.effect?.updateTiming({ easing: "linear" }), this.animation.onfinish = null, e && s() ? (this.animation.timeline = e, n && (this.animation.rangeStart = n), r && (this.animation.rangeEnd = r), t) : i(this);
	}
};
//#endregion
export { u as NativeAnimation };
