// motion-dom/dist/es/animation/NativeAnimationExtended.js
import { clamp as e } from "../../../../motion-utils/dist/es/clamp.js";
import { time as t } from "../frameloop/sync-time.js";
import { replaceTransitionType as n } from "./utils/replace-transition-type.js";
import { JSAnimation as r } from "./JSAnimation.js";
import { setStyle as i } from "../render/dom/style-set.js";
import { NativeAnimation as a } from "./NativeAnimation.js";
import { replaceStringEasing as o } from "./waapi/utils/unsupported-easing.js";
//#region node_modules/motion-dom/dist/es/animation/NativeAnimationExtended.mjs
var s = 10, c = class extends a {
	constructor(e) {
		o(e), n(e), super(e), e.startTime !== void 0 && e.autoplay !== !1 && (this.startTime = e.startTime), this.options = e;
	}
	updateMotionValue(n) {
		let { motionValue: a, onUpdate: o, onComplete: c, element: l, ...u } = this.options;
		if (!a) return;
		if (n !== void 0) {
			a.set(n);
			return;
		}
		let d = new r({
			...u,
			autoplay: !1
		}), f = Math.max(s, t.now() - this.startTime), p = e(0, s, f - s), m = d.sample(f).value, { name: h } = this.options;
		l && h && i(l, h, m), a.setWithVelocity(d.sample(Math.max(0, f - p)).value, m, p), d.stop();
	}
};
//#endregion
export { c as NativeAnimationExtended };
