// motion-dom/dist/es/value/types/maps/number.js
import { alpha as e } from "../numbers/index.js";
import { px as t } from "../numbers/units.js";
import { int as n } from "../int.js";
import { transformValueTypes as r } from "./transform.js";
//#region node_modules/motion-dom/dist/es/value/types/maps/number.mjs
var i = {
	borderWidth: t,
	borderTopWidth: t,
	borderRightWidth: t,
	borderBottomWidth: t,
	borderLeftWidth: t,
	borderRadius: t,
	borderTopLeftRadius: t,
	borderTopRightRadius: t,
	borderBottomRightRadius: t,
	borderBottomLeftRadius: t,
	width: t,
	maxWidth: t,
	height: t,
	maxHeight: t,
	top: t,
	right: t,
	bottom: t,
	left: t,
	inset: t,
	insetBlock: t,
	insetBlockStart: t,
	insetBlockEnd: t,
	insetInline: t,
	insetInlineStart: t,
	insetInlineEnd: t,
	padding: t,
	paddingTop: t,
	paddingRight: t,
	paddingBottom: t,
	paddingLeft: t,
	paddingBlock: t,
	paddingBlockStart: t,
	paddingBlockEnd: t,
	paddingInline: t,
	paddingInlineStart: t,
	paddingInlineEnd: t,
	margin: t,
	marginTop: t,
	marginRight: t,
	marginBottom: t,
	marginLeft: t,
	marginBlock: t,
	marginBlockStart: t,
	marginBlockEnd: t,
	marginInline: t,
	marginInlineStart: t,
	marginInlineEnd: t,
	fontSize: t,
	backgroundPositionX: t,
	backgroundPositionY: t,
	...r,
	zIndex: n,
	fillOpacity: e,
	strokeOpacity: e,
	numOctaves: n
};
//#endregion
export { i as numberValueTypes };
