// @base-ui/react/esm/internals/useRenderElement.js
import { mergeObjects as e } from "../../../utils/esm/mergeObjects.js";
import { mergeClassNames as t, mergeProps as n, mergePropsN as r } from "../merge-props/mergeProps.js";
import i from "../../../utils/esm/formatErrorMessage.js";
import { useMergedRefs as a, useMergedRefsN as o } from "../../../utils/esm/useMergedRefs.js";
import { getReactElementRef as s } from "../../../utils/esm/getReactElementRef.js";
import { warn as c } from "../../../utils/esm/warn.js";
import { EMPTY_OBJECT as l } from "../../../utils/esm/empty.js";
import { getStateAttributesProps as u } from "./getStateAttributesProps.js";
import { resolveClassName as d } from "../utils/resolveClassName.js";
import { resolveStyle as f } from "../utils/resolveStyle.js";
import * as p from "react";
import { createElement as m } from "react";
//#region node_modules/@base-ui/react/esm/internals/useRenderElement.js
function h(e, t, n = {}) {
	let r = t.render, i = g(t, n);
	return n.enabled === !1 ? null : x(e, r, i, n.state ?? l);
}
function g(n, r = {}) {
	let { className: i, style: c, render: p } = n, { state: m = l, ref: h, props: g, stateAttributesMapping: v, enabled: y = !0 } = r, b = y ? d(i, m) : void 0, x = y ? f(c, m) : void 0, S = y ? u(m, v) : l, C = y && g ? _(g) : void 0, w = y ? e(S, C) ?? {} : l;
	return typeof document < "u" && (y ? Array.isArray(h) ? w.ref = o([
		w.ref,
		s(p),
		...h
	]) : w.ref = a(w.ref, s(p), h) : a(null, null)), y ? (b !== void 0 && (w.className = t(w.className, b)), x !== void 0 && (w.style = e(w.style, x)), w) : l;
}
function _(e) {
	return Array.isArray(e) ? r(e) : n(void 0, e);
}
var v = Symbol.for("react.lazy"), y = /^[A-Z][A-Za-z0-9$]*$/, b = /[a-z]/;
function x(e, t, r, a) {
	if (t) {
		if (typeof t == "function") return process.env.NODE_ENV !== "production" && S(t), t(r, a);
		let e = n(r, t.props);
		e.ref = r.ref;
		let i = t;
		if (i?.$$typeof === v && (i = p.Children.toArray(t)[0]), process.env.NODE_ENV !== "production" && !/*#__PURE__*/ p.isValidElement(i)) throw Error([
			"Base UI: The `render` prop was provided an invalid React element as `React.isValidElement(render)` is `false`.",
			"A valid React element must be provided to the `render` prop because it is cloned with props to replace the default element.",
			"https://base-ui.com/r/invalid-render-prop"
		].join("\n"));
		return /*#__PURE__*/ p.cloneElement(i, e);
	}
	if (e && typeof e == "string") return C(e, r);
	throw Error(process.env.NODE_ENV === "production" ? i(8) : "Base UI: Render element or function are not defined.");
}
function S(e) {
	let t = e.name;
	t.length !== 0 && y.test(t) && b.test(t) && c(`The \`render\` prop received a function named \`${t}\` that starts with an uppercase letter.`, "This usually means a React component was passed directly as `render={Component}`.", "Base UI calls `render` as a plain function, which can break the Rules of Hooks during reconciliation.", "If this is an intentional render callback, rename it to start with a lowercase letter.", "Use `render={<Component />}` or `render={(props) => <Component {...props} />}` instead.", "https://base-ui.com/r/invalid-render-prop");
}
function C(e, t) {
	return e === "button" ? /*#__PURE__*/ m("button", {
		type: "button",
		...t,
		key: t.key
	}) : e === "img" ? /*#__PURE__*/ m("img", {
		alt: "",
		...t,
		key: t.key
	}) : /*#__PURE__*/ p.createElement(e, t);
}
//#endregion
export { h as useRenderElement };
