// framer-motion/dist/es/gestures/drag/utils/constraints.js
import { clamp as e } from "../../../../../../motion-utils/dist/es/clamp.js";
import { progress as t } from "../../../../../../motion-utils/dist/es/progress.js";
import { mixNumber as n } from "../../../../../../motion-dom/dist/es/utils/mix/number.js";
import { calcLength as r } from "../../../../../../motion-dom/dist/es/projection/geometry/delta-calc.js";
//#region node_modules/framer-motion/dist/es/gestures/drag/utils/constraints.mjs
function i(e, { min: t, max: r }, i) {
	return t !== void 0 && e < t ? e = i ? n(t, e, i.min) : Math.max(e, t) : r !== void 0 && e > r && (e = i ? n(r, e, i.max) : Math.min(e, r)), e;
}
function a(e, t, n) {
	return {
		min: t === void 0 ? void 0 : e.min + t,
		max: n === void 0 ? void 0 : e.max + n - (e.max - e.min)
	};
}
function o(e, { top: t, left: n, bottom: r, right: i }) {
	return {
		x: a(e.x, n, i),
		y: a(e.y, t, r)
	};
}
function s(e, t) {
	let n = t.min - e.min, r = t.max - e.max;
	return t.max - t.min < e.max - e.min && ([n, r] = [r, n]), {
		min: n,
		max: r
	};
}
function c(e, t) {
	return {
		x: s(e.x, t.x),
		y: s(e.y, t.y)
	};
}
function l(n, i) {
	let a = .5, o = r(n), s = r(i);
	return s > o ? a = t(i.min, i.max - o, n.min) : o > s && (a = t(n.min, n.max - s, i.min)), e(0, 1, a);
}
function u(e, t) {
	let n = {};
	return t.min !== void 0 && (n.min = t.min - e.min), t.max !== void 0 && (n.max = t.max - e.min), n;
}
var d = .35;
function f(e = d) {
	return e === !1 ? e = 0 : e === !0 && (e = d), {
		x: p(e, "left", "right"),
		y: p(e, "top", "bottom")
	};
}
function p(e, t, n) {
	return {
		min: m(e, t),
		max: m(e, n)
	};
}
function m(e, t) {
	return typeof e == "number" ? e : e[t] || 0;
}
//#endregion
export { i as applyConstraints, l as calcOrigin, o as calcRelativeConstraints, c as calcViewportConstraints, d as defaultElastic, u as rebaseAxisConstraints, f as resolveDragElastic };
