// @base-ui/react/esm/internals/use-button/useButton.js
"use client";
import { isHTMLElement as e } from "../../../../../@floating-ui/utils/dist/floating-ui.utils.dom.js";
import { SafeReact as t } from "../../../../utils/esm/safeReact.js";
import { useStableCallback as n } from "../../../../utils/esm/useStableCallback.js";
import { error as r } from "../../../../utils/esm/error.js";
import { useIsoLayoutEffect as i } from "../../../../utils/esm/useIsoLayoutEffect.js";
import { makeEventPreventable as a, mergeProps as o } from "../../merge-props/mergeProps.js";
import { useCompositeRootContext as s } from "../composite/root/CompositeRootContext.js";
import { useFocusableWhenDisabled as c } from "../../utils/useFocusableWhenDisabled.js";
import * as l from "react";
//#region node_modules/@base-ui/react/esm/internals/use-button/useButton.js
function u(e = {}) {
	let { disabled: u = !1, focusableWhenDisabled: p, tabIndex: m = 0, native: h = !0, composite: g } = e, _ = l.useRef(null), v = s(!0), y = g ?? v !== void 0, { props: b } = c({
		focusableWhenDisabled: p,
		disabled: u,
		composite: y,
		tabIndex: m,
		isNativeButton: h
	});
	process.env.NODE_ENV !== "production" && l.useEffect(() => {
		if (!_.current) return;
		let e = d(_.current);
		h ? e || r(`A component that acts as a button expected a native <button> because the \`nativeButton\` prop is true. Rendering a non-<button> removes native button semantics, which can impact forms and accessibility. Use a real <button> in the \`render\` prop, or set \`nativeButton\` to \`false\`.${t.captureOwnerStack?.() || ""}`) : e && r(`A component that acts as a button expected a non-<button> because the \`nativeButton\` prop is false. Rendering a <button> keeps native behavior while Base UI applies non-native attributes and handlers, which can add unintended extra attributes (such as \`role\` or \`aria-disabled\`). Use a non-<button> in the \`render\` prop, or set \`nativeButton\` to \`true\`.${t.captureOwnerStack?.() || ""}`);
	}, [h]);
	let x = l.useCallback(() => {
		let e = _.current;
		d(e) && y && u && b.disabled === void 0 && e.disabled && (e.disabled = !1);
	}, [
		u,
		b.disabled,
		y
	]);
	return i(x, [x]), {
		getButtonProps: l.useCallback((e = {}) => {
			let { onClick: t, onMouseDown: n, onKeyUp: r, onKeyDown: i, onPointerDown: s, ...c } = e;
			return o({
				onClick(e) {
					if (u) {
						e.preventDefault();
						return;
					}
					t?.(e);
				},
				onMouseDown(e) {
					u || n?.(e);
				},
				onKeyDown(e) {
					if (u || (a(e), i?.(e), e.baseUIHandlerPrevented)) return;
					let n = e.target === e.currentTarget, r = e.currentTarget, o = d(r), s = !h && f(r), c = n && (h ? o : !s), l = e.key === "Enter", p = e.key === " ", m = r.getAttribute("role"), g = m?.startsWith("menuitem") || m === "option" || m === "gridcell";
					if (n && y && p) {
						if (e.defaultPrevented && g) return;
						e.preventDefault(), s || h && o ? (r.click(), e.preventBaseUIHandler()) : c && (t?.(e), e.preventBaseUIHandler());
						return;
					}
					c && (!h && (p || l) && e.preventDefault(), !h && l && t?.(e));
				},
				onKeyUp(e) {
					if (!u) {
						if (a(e), r?.(e), e.target === e.currentTarget && h && y && d(e.currentTarget) && e.key === " ") {
							e.preventDefault();
							return;
						}
						e.baseUIHandlerPrevented || e.target === e.currentTarget && !h && !y && e.key === " " && t?.(e);
					}
				},
				onPointerDown(e) {
					if (u) {
						e.preventDefault();
						return;
					}
					s?.(e);
				}
			}, h ? { type: "button" } : { role: "button" }, b, c);
		}, [
			u,
			b,
			y,
			h
		]),
		buttonRef: n((e) => {
			_.current = e, x();
		})
	};
}
function d(t) {
	return e(t) && t.tagName === "BUTTON";
}
function f(e) {
	return !!(e?.tagName === "A" && e?.href);
}
//#endregion
export { u as useButton };
