// framer-motion/dist/es/motion/index.js
"use client";
import { LayoutGroupContext as e } from "../context/LayoutGroupContext.js";
import { invariant as t, warning as n } from "../../../../motion-utils/dist/es/errors.js";
import { MotionConfigContext as r } from "../context/MotionConfigContext.js";
import { LazyContext as i } from "../context/LazyContext.js";
import { getInitializedFeatureDefinitions as a } from "./features/definitions.js";
import { loadFeatures as o } from "./features/load-features.js";
import { MotionContext as s } from "../context/MotionContext/index.js";
import { useCreateMotionContext as c } from "../context/MotionContext/create.js";
import { isSVGComponent as l } from "../render/dom/utils/is-svg-component.js";
import { useRender as u } from "../render/dom/use-render.js";
import { useHTMLVisualState as d } from "../render/html/use-html-visual-state.js";
import { useSVGVisualState as f } from "../render/svg/use-svg-visual-state.js";
import { motionComponentSymbol as p } from "./utils/symbol.js";
import { useMotionRef as m } from "./utils/use-motion-ref.js";
import { useVisualElement as h } from "./utils/use-visual-element.js";
import { jsx as g, jsxs as _ } from "react/jsx-runtime";
import { forwardRef as v, useContext as y } from "react";
//#region node_modules/framer-motion/dist/es/motion/index.mjs
function b(e, { forwardMotionProps: t = !1, type: n } = {}, i, a) {
	i && o(i);
	let b = n ? n === "svg" : l(e), w = b ? f : d;
	function T(n, o) {
		let l, d = {
			...y(r),
			...n,
			layoutId: x(n)
		}, { isStatic: f } = d, p = c(n), v = w(n, f);
		if (!f && typeof window < "u") {
			S(d, i);
			let t = C(d);
			l = t.MeasureLayout, p.visualElement = h(e, v, d, a, t.ProjectionNode, b);
		}
		return _(s.Provider, {
			value: p,
			children: [l && p.visualElement ? g(l, {
				visualElement: p.visualElement,
				...d
			}) : null, u(e, n, m(v, p.visualElement, o), v, f, t, b)]
		});
	}
	T.displayName = `motion.${typeof e == "string" ? e : `create(${e.displayName ?? e.name ?? ""})`}`;
	let E = v(T);
	return E[p] = e, E;
}
function x({ layoutId: t }) {
	let n = y(e).id;
	return n && t !== void 0 ? n + "-" + t : t;
}
function S(e, r) {
	let a = y(i).strict;
	if (process.env.NODE_ENV !== "production" && r && a) {
		let r = "You have rendered a `motion` component within a `LazyMotion` component. This will break tree shaking. Import and render a `m` component instead.";
		e.ignoreStrict ? n(!1, r, "lazy-strict-mode") : t(!1, r, "lazy-strict-mode");
	}
}
function C(e) {
	let { drag: t, layout: n } = a();
	if (!t && !n) return {};
	let r = {
		...t,
		...n
	};
	return {
		MeasureLayout: t?.isEnabled(e) || n?.isEnabled(e) ? r.MeasureLayout : void 0,
		ProjectionNode: r.ProjectionNode
	};
}
//#endregion
export { b as createMotionComponent };
