// motion-dom/dist/es/projection/shared/stack.js
import { addUniqueItem as e, removeItem as t } from "../../../../../motion-utils/dist/es/array.js";
//#region node_modules/motion-dom/dist/es/projection/shared/stack.mjs
var n = class {
	constructor() {
		this.members = [];
	}
	add(n) {
		e(this.members, n);
		for (let e = this.members.length - 1; e >= 0; e--) {
			let r = this.members[e];
			if (r === n || r === this.lead || r === this.prevLead) continue;
			let i = r.instance;
			(!i || i.isConnected === !1) && !r.snapshot && (t(this.members, r), r.unmount());
		}
		n.scheduleRender();
	}
	remove(e) {
		if (t(this.members, e), e === this.prevLead && (this.prevLead = void 0), e === this.lead) {
			let e = this.members[this.members.length - 1];
			e && this.promote(e);
		}
	}
	relegate(e) {
		for (let t = this.members.indexOf(e) - 1; t >= 0; t--) {
			let e = this.members[t];
			if (e.isPresent !== !1 && e.instance?.isConnected !== !1) return this.promote(e), !0;
		}
		return !1;
	}
	promote(e, t) {
		let n = this.lead;
		if (e !== n && (this.prevLead = n, this.lead = e, e.show(), n)) {
			n.updateSnapshot(), e.scheduleRender();
			let { layoutDependency: r } = n.options, { layoutDependency: i } = e.options;
			(r === void 0 || r !== i) && (e.resumeFrom = n, t && (n.preserveOpacity = !0), n.snapshot && (e.snapshot = n.snapshot, e.snapshot.latestValues = n.animationValues || n.latestValues), e.root?.isUpdating && (e.isLayoutDirty = !0)), e.options.crossfade === !1 && n.hide();
		}
	}
	exitAnimationComplete() {
		this.members.forEach((e) => {
			e.options.onExitComplete?.(), e.resumingFrom?.options.onExitComplete?.();
		});
	}
	scheduleRender() {
		this.members.forEach((e) => e.instance && e.scheduleRender(!1));
	}
	removeLeadSnapshot() {
		this.lead?.snapshot && (this.lead.snapshot = void 0);
	}
};
//#endregion
export { n as NodeStack };
