// motion-dom/dist/es/value/index.js
import { SubscriptionManager as e } from "../../../../motion-utils/dist/es/subscription-manager.js";
import { velocityPerSecond as t } from "../../../../motion-utils/dist/es/velocity-per-second.js";
import { warnOnce as n } from "../../../../motion-utils/dist/es/warn-once.js";
import { frame as r } from "../frameloop/frame.js";
import { time as i } from "../frameloop/sync-time.js";
//#region node_modules/motion-dom/dist/es/value/index.mjs
var a = 30, o = (e) => !isNaN(parseFloat(e)), s = { current: void 0 }, c = class {
	constructor(e, t = {}) {
		this.canTrackVelocity = null, this.events = {}, this.updateAndNotify = (e) => {
			let t = i.now();
			if (this.updatedAt !== t && this.setPrevFrameValue(), this.prev = this.current, this.setCurrent(e), this.current !== this.prev && (this.events.change?.notify(this.current), this.dependents)) for (let e of this.dependents) e.dirty();
		}, this.hasAnimated = !1, this.setCurrent(e), this.owner = t.owner;
	}
	setCurrent(e) {
		this.current = e, this.updatedAt = i.now(), this.canTrackVelocity === null && e !== void 0 && (this.canTrackVelocity = o(this.current));
	}
	setPrevFrameValue(e = this.current) {
		this.prevFrameValue = e, this.prevUpdatedAt = this.updatedAt;
	}
	onChange(e) {
		return process.env.NODE_ENV !== "production" && n(!1, "value.onChange(callback) is deprecated. Switch to value.on(\"change\", callback)."), this.on("change", e);
	}
	on(t, n) {
		this.events[t] || (this.events[t] = new e());
		let i = this.events[t].add(n);
		return t === "change" ? () => {
			i(), r.read(() => {
				this.events.change.getSize() || this.stop();
			});
		} : i;
	}
	clearListeners() {
		for (let e in this.events) this.events[e].clear();
	}
	attach(e, t) {
		this.passiveEffect = e, this.stopPassiveEffect = t;
	}
	set(e) {
		this.passiveEffect ? this.passiveEffect(e, this.updateAndNotify) : this.updateAndNotify(e);
	}
	setWithVelocity(e, t, n) {
		this.set(t), this.prev = void 0, this.prevFrameValue = e, this.prevUpdatedAt = this.updatedAt - n;
	}
	jump(e, t = !0) {
		this.updateAndNotify(e), this.prev = e, this.prevUpdatedAt = this.prevFrameValue = void 0, t && this.stop(), this.stopPassiveEffect && this.stopPassiveEffect();
	}
	dirty() {
		this.events.change?.notify(this.current);
	}
	addDependent(e) {
		this.dependents ||= /* @__PURE__ */ new Set(), this.dependents.add(e);
	}
	removeDependent(e) {
		this.dependents && this.dependents.delete(e);
	}
	get() {
		return s.current && s.current.push(this), this.current;
	}
	getPrevious() {
		return this.prev;
	}
	getVelocity() {
		let e = i.now();
		if (!this.canTrackVelocity || this.prevFrameValue === void 0 || e - this.updatedAt > a) return 0;
		let n = Math.min(this.updatedAt - this.prevUpdatedAt, a);
		return t(parseFloat(this.current) - parseFloat(this.prevFrameValue), n);
	}
	start(e) {
		return this.stop(), new Promise((t) => {
			this.hasAnimated = !0, this.animation = e(t), this.events.animationStart && this.events.animationStart.notify();
		}).then(() => {
			this.events.animationComplete && this.events.animationComplete.notify(), this.clearAnimation();
		});
	}
	stop() {
		this.animation && (this.animation.stop(), this.events.animationCancel && this.events.animationCancel.notify()), this.clearAnimation();
	}
	isAnimating() {
		return !!this.animation;
	}
	clearAnimation() {
		delete this.animation;
	}
	destroy() {
		this.dependents?.clear(), this.events.destroy?.notify(), this.clearListeners(), this.stop(), this.stopPassiveEffect && this.stopPassiveEffect();
	}
};
function l(e, t) {
	return new c(e, t);
}
//#endregion
export { l as motionValue };
