// framer-motion/dist/es/motion/features/viewport/index.js
import { Feature as e } from "../../../../../../motion-dom/dist/es/render/Feature.js";
import { observeIntersection as t } from "./observers.js";
//#region node_modules/framer-motion/dist/es/motion/features/viewport/index.mjs
var n = {
	some: 0,
	all: 1
}, r = class extends e {
	constructor() {
		super(...arguments), this.hasEnteredView = !1, this.isInView = !1;
	}
	startObserver() {
		this.stopObserver?.();
		let { viewport: e = {} } = this.node.getProps(), { root: r, margin: i, amount: a = "some", once: o } = e, s = {
			root: r ? r.current : void 0,
			rootMargin: i,
			threshold: typeof a == "number" ? a : n[a]
		}, c = (e) => {
			let { isIntersecting: t } = e;
			if (this.isInView === t || (this.isInView = t, o && !t && this.hasEnteredView)) return;
			t && (this.hasEnteredView = !0), this.node.animationState && this.node.animationState.setActive("whileInView", t);
			let { onViewportEnter: n, onViewportLeave: r } = this.node.getProps(), i = t ? n : r;
			i && i(e);
		};
		this.stopObserver = t(this.node.current, s, c);
	}
	mount() {
		this.startObserver();
	}
	update() {
		if (typeof IntersectionObserver > "u") return;
		let { props: e, prevProps: t } = this.node;
		[
			"amount",
			"margin",
			"root"
		].some(i(e, t)) && this.startObserver();
	}
	unmount() {
		this.stopObserver?.(), this.hasEnteredView = !1, this.isInView = !1;
	}
};
function i({ viewport: e = {} }, { viewport: t = {} } = {}) {
	return (n) => e[n] !== t[n];
}
//#endregion
export { r as InViewFeature };
