// framer-motion/dist/es/components/AnimatePresence/index.js
"use client";
import { LayoutGroupContext as e } from "../../context/LayoutGroupContext.js";
import { useConstant as t } from "../../utils/use-constant.js";
import { useIsomorphicLayoutEffect as n } from "../../utils/use-isomorphic-effect.js";
import { PresenceChild as r } from "./PresenceChild.js";
import { usePresence as i } from "./use-presence.js";
import { getChildKey as a, onlyElements as o } from "./utils.js";
import { Fragment as s, jsx as c } from "react/jsx-runtime";
import { useContext as l, useMemo as u, useRef as d, useState as f } from "react";
//#region node_modules/framer-motion/dist/es/components/AnimatePresence/index.mjs
var p = ({ children: p, custom: m, initial: h = !0, onExitComplete: g, presenceAffectsLayout: _ = !0, mode: v = "sync", propagate: y = !1, anchorX: b = "left", anchorY: x = "top", root: S }) => {
	let [C, w] = i(y), T = u(() => o(p), [p]), E = y && !C ? [] : T.map(a), D = d(!0), O = d(T), k = t(() => /* @__PURE__ */ new Map()), A = d(/* @__PURE__ */ new Set()), [j, M] = f(T), [N, P] = f(T);
	n(() => {
		D.current = !1, O.current = T;
		for (let e = 0; e < N.length; e++) {
			let t = a(N[e]);
			E.includes(t) ? (k.delete(t), A.current.delete(t)) : k.get(t) !== !0 && k.set(t, !1);
		}
	}, [
		N,
		E.length,
		E.join("-")
	]);
	let F = [];
	if (T !== j) {
		let e = [...T];
		for (let t = 0; t < N.length; t++) {
			let n = N[t], r = a(n);
			E.includes(r) || (e.splice(t, 0, n), F.push(n));
		}
		return v === "wait" && F.length && (e = F), P(o(e)), M(T), null;
	}
	process.env.NODE_ENV !== "production" && v === "wait" && N.length > 1 && console.warn("You're attempting to animate multiple children within AnimatePresence, but its mode is set to \"wait\". This will lead to odd visual behaviour.");
	let { forceRender: I } = l(e);
	return c(s, { children: N.map((e) => {
		let t = a(e), n = y && !C ? !1 : T === N || E.includes(t);
		return c(r, {
			isPresent: n,
			initial: !D.current || h ? void 0 : !1,
			custom: m,
			presenceAffectsLayout: _,
			mode: v,
			root: S,
			onExitComplete: n ? void 0 : () => {
				if (A.current.has(t)) return;
				if (k.has(t)) A.current.add(t), k.set(t, !0);
				else return;
				let e = !0;
				k.forEach((t) => {
					t || (e = !1);
				}), e && (I?.(), P(O.current), y && w?.(), g && g());
			},
			anchorX: b,
			anchorY: x,
			children: e
		}, t);
	}) });
};
//#endregion
export { p as AnimatePresence };
