// framer-motion/dist/es/components/AnimatePresence/PopChild.js
"use client";
import { isHTMLElement as e } from "../../../../../motion-dom/dist/es/utils/is-html-element.js";
import { MotionConfigContext as t } from "../../context/MotionConfigContext.js";
import { useComposedRefs as n } from "../../utils/use-composed-ref.js";
import { jsx as r } from "react/jsx-runtime";
import * as i from "react";
import { useContext as a, useId as o, useInsertionEffect as s, useRef as c } from "react";
//#region node_modules/framer-motion/dist/es/components/AnimatePresence/PopChild.mjs
var l = class extends i.Component {
	getSnapshotBeforeUpdate(t) {
		let n = this.props.childRef.current;
		if (e(n) && t.isPresent && !this.props.isPresent && this.props.pop !== !1) {
			let t = n.offsetParent, r = e(t) && t.offsetWidth || 0, i = e(t) && t.offsetHeight || 0, a = getComputedStyle(n), o = this.props.sizeRef.current;
			o.height = parseFloat(a.height), o.width = parseFloat(a.width), o.top = n.offsetTop, o.left = n.offsetLeft, o.right = r - o.width - o.left, o.bottom = i - o.height - o.top, o.direction = a.direction;
		}
		return null;
	}
	componentDidUpdate() {}
	render() {
		return this.props.children;
	}
};
function u({ children: e, isPresent: u, anchorX: d, anchorY: f, root: p, pop: m }) {
	let h = o(), g = c(null), _ = c({
		width: 0,
		height: 0,
		top: 0,
		left: 0,
		right: 0,
		bottom: 0,
		direction: "ltr"
	}), { nonce: v } = a(t), y = n(g, e.props?.ref ?? e?.ref);
	return s(() => {
		let { width: e, height: t, top: n, left: r, right: i, bottom: a, direction: o } = _.current;
		if (u || m === !1 || !g.current || !e || !t) return;
		let s = o === "rtl", c = d === "left" ? s ? `right: ${i}` : `left: ${r}` : s ? `left: ${r}` : `right: ${i}`, l = f === "bottom" ? `bottom: ${a}` : `top: ${n}`;
		g.current.dataset.motionPopId = h;
		let y = document.createElement("style");
		v && (y.nonce = v);
		let b = p ?? document.head;
		return b.appendChild(y), y.sheet && y.sheet.insertRule(`
          [data-motion-pop-id="${h}"] {
            position: absolute !important;
            width: ${e}px !important;
            height: ${t}px !important;
            ${c}px !important;
            ${l}px !important;
          }
        `), () => {
			g.current?.removeAttribute("data-motion-pop-id"), b.contains(y) && b.removeChild(y);
		};
	}, [u]), r(l, {
		isPresent: u,
		childRef: g,
		sizeRef: _,
		pop: m,
		children: m === !1 ? e : i.cloneElement(e, { ref: y })
	});
}
//#endregion
export { u as PopChild };
