// framer-motion/dist/es/components/AnimatePresence/PresenceChild.js
"use client";
import { useConstant as e } from "../../utils/use-constant.js";
import { PresenceContext as t } from "../../context/PresenceContext.js";
import { PopChild as n } from "./PopChild.js";
import { jsx as r } from "react/jsx-runtime";
import * as i from "react";
import { useId as a, useMemo as o } from "react";
//#region node_modules/framer-motion/dist/es/components/AnimatePresence/PresenceChild.mjs
var s = ({ children: s, initial: l, isPresent: u, onExitComplete: d, custom: f, presenceAffectsLayout: p, mode: m, anchorX: h, anchorY: g, root: _ }) => {
	let v = e(c), y = a(), b = !0, x = o(() => (b = !1, {
		id: y,
		initial: l,
		isPresent: u,
		custom: f,
		onExitComplete: (e) => {
			v.set(e, !0);
			for (let e of v.values()) if (!e) return;
			d && d();
		},
		register: (e) => (v.set(e, !1), () => v.delete(e))
	}), [
		u,
		v,
		d
	]);
	return p && b && (x = { ...x }), o(() => {
		v.forEach((e, t) => v.set(t, !1));
	}, [u]), i.useEffect(() => {
		!u && !v.size && d && d();
	}, [u]), s = r(n, {
		pop: m === "popLayout",
		isPresent: u,
		anchorX: h,
		anchorY: g,
		root: _,
		children: s
	}), r(t.Provider, {
		value: x,
		children: s
	});
};
function c() {
	return /* @__PURE__ */ new Map();
}
//#endregion
export { s as PresenceChild };
