// motion-dom/dist/es/projection/node/create-projection-node.js
import { clamp as e } from "../../../../../motion-utils/dist/es/clamp.js";
import { noop as t } from "../../../../../motion-utils/dist/es/noop.js";
import { SubscriptionManager as n } from "../../../../../motion-utils/dist/es/subscription-manager.js";
import { statsBuffer as r } from "../../stats/buffer.js";
import { cancelFrame as i, frame as a, frameData as o, frameSteps as s } from "../../frameloop/frame.js";
import { time as c } from "../../frameloop/sync-time.js";
import { activeAnimations as l } from "../../stats/animation-count.js";
import { mixNumber as u } from "../../utils/mix/number.js";
import { motionValue as d } from "../../value/index.js";
import { getValueTransition as f } from "../../animation/utils/get-value-transition.js";
import { getOptimisedAppearId as p } from "../../animation/optimized-appear/get-appear-id.js";
import { microtask as m } from "../../frameloop/microtask.js";
import { isSVGElement as ee } from "../../utils/is-svg-element.js";
import { isSVGSVGElement as te } from "../../utils/is-svg-svg-element.js";
import { createBox as h, createDelta as g } from "../geometry/models.js";
import { has2DTranslate as _, hasScale as v, hasTransform as y } from "../utils/has-transform.js";
import { applyBoxDelta as ne, applyTreeDeltas as re, transformBox as b, translateAxis as x } from "../geometry/delta-apply.js";
import { scaleCorrectors as S } from "../styles/scale-correction.js";
import { copyAxisDeltaInto as C, copyAxisInto as w, copyBoxInto as T } from "../geometry/copy.js";
import { calcBoxDelta as E, calcLength as D, calcRelativeBox as ie, calcRelativePosition as O, isNear as k } from "../geometry/delta-calc.js";
import { removeBoxTransforms as A } from "../geometry/delta-remove.js";
import { aspectRatio as j, axisDeltaEquals as M, boxEquals as ae, boxEqualsRounded as N, isDeltaZero as P } from "../geometry/utils.js";
import { eachAxis as F } from "../utils/each-axis.js";
import { buildProjectionTransform as oe } from "../styles/transform.js";
import { mixValues as se } from "../animation/mix-values.js";
import { animateSingleValue as ce } from "../../animation/animate/single-value.js";
import { FlatTree as le } from "../utils/flat-tree.js";
import { delay as ue } from "../../utils/delay.js";
import { resolveMotionValue as I } from "../../value/utils/resolve-motion-value.js";
import { NodeStack as de } from "../shared/stack.js";
import { globalProjectionState as L } from "./state.js";
//#region node_modules/motion-dom/dist/es/projection/node/create-projection-node.mjs
var R = {
	nodes: 0,
	calculatedTargetDeltas: 0,
	calculatedProjections: 0
}, z = [
	"",
	"X",
	"Y",
	"Z"
], B = 1e3, V = 0;
function H(e, t, n, r) {
	let { latestValues: i } = t;
	i[e] && (n[e] = i[e], t.setStaticValue(e, 0), r && (r[e] = 0));
}
function U(e) {
	if (e.hasCheckedOptimisedAppear = !0, e.root === e) return;
	let { visualElement: t } = e.options;
	if (!t) return;
	let n = p(t);
	if (window.MotionHasOptimisedAnimation(n, "transform")) {
		let { layout: t, layoutId: r } = e.options;
		window.MotionCancelOptimisedAnimation(n, "transform", a, !(t || r));
	}
	let { parent: r } = e;
	r && !r.hasCheckedOptimisedAppear && U(r);
}
function fe({ attachResizeListener: t, defaultParent: p, measureScroll: w, checkIsScrollRoot: k, resetTransform: j }) {
	return class {
		constructor(e = {}, t = p?.()) {
			this.id = V++, this.animationId = 0, this.animationCommitId = 0, this.children = /* @__PURE__ */ new Set(), this.options = {}, this.isTreeAnimating = !1, this.isAnimationBlocked = !1, this.isLayoutDirty = !1, this.isProjectionDirty = !1, this.isSharedProjectionDirty = !1, this.isTransformDirty = !1, this.updateManuallyBlocked = !1, this.updateBlockedByResize = !1, this.isUpdating = !1, this.isSVG = !1, this.needsReset = !1, this.shouldResetTransform = !1, this.hasCheckedOptimisedAppear = !1, this.treeScale = {
				x: 1,
				y: 1
			}, this.eventHandlers = /* @__PURE__ */ new Map(), this.hasTreeAnimated = !1, this.layoutVersion = 0, this.updateScheduled = !1, this.scheduleUpdate = () => this.update(), this.projectionUpdateScheduled = !1, this.checkUpdateFailed = () => {
				this.isUpdating && (this.isUpdating = !1, this.clearAllSnapshots());
			}, this.updateProjection = () => {
				this.projectionUpdateScheduled = !1, r.value && (R.nodes = R.calculatedTargetDeltas = R.calculatedProjections = 0), this.nodes.forEach(he), this.nodes.forEach(xe), this.nodes.forEach(Se), this.nodes.forEach(ge), r.addProjectionMetrics && r.addProjectionMetrics(R);
			}, this.resolvedRelativeTargetAt = 0, this.linkedParentVersion = 0, this.hasProjected = !1, this.isVisible = !0, this.animationProgress = 0, this.sharedNodes = /* @__PURE__ */ new Map(), this.latestValues = e, this.root = t ? t.root || t : this, this.path = t ? [...t.path, t] : [], this.parent = t, this.depth = t ? t.depth + 1 : 0;
			for (let e = 0; e < this.path.length; e++) this.path[e].shouldResetTransform = !0;
			this.root === this && (this.nodes = new le());
		}
		addEventListener(e, t) {
			return this.eventHandlers.has(e) || this.eventHandlers.set(e, new n()), this.eventHandlers.get(e).add(t);
		}
		notifyListeners(e, ...t) {
			let n = this.eventHandlers.get(e);
			n && n.notify(...t);
		}
		hasListeners(e) {
			return this.eventHandlers.has(e);
		}
		mount(e) {
			if (this.instance) return;
			this.isSVG = ee(e) && !te(e), this.instance = e;
			let { layoutId: n, layout: r, visualElement: i } = this.options;
			if (i && !i.current && i.mount(e), this.root.nodes.add(this), this.parent && this.parent.children.add(this), this.root.hasTreeAnimated && (r || n) && (this.isLayoutDirty = !0), t) {
				let n, r = 0, i = () => this.root.updateBlockedByResize = !1;
				a.read(() => {
					r = window.innerWidth;
				}), t(e, () => {
					let e = window.innerWidth;
					e !== r && (r = e, this.root.updateBlockedByResize = !0, n && n(), n = ue(i, 250), L.hasAnimatedSinceResize && (L.hasAnimatedSinceResize = !1, this.nodes.forEach(K)));
				});
			}
			n && this.root.registerSharedNode(n, this), this.options.animate !== !1 && i && (n || r) && this.addEventListener("didUpdate", ({ delta: e, hasLayoutChanged: t, hasRelativeLayoutChanged: n, layout: r }) => {
				if (this.isTreeAnimationBlocked()) {
					this.target = void 0, this.relativeTarget = void 0;
					return;
				}
				let a = this.options.transition || i.getDefaultTransition() || Ee, { onLayoutAnimationStart: o, onLayoutAnimationComplete: s } = i.getProps(), c = !this.targetLayout || !N(this.targetLayout, r), l = !t && n;
				if (this.options.layoutRoot || this.resumeFrom || l || t && (c || !this.currentAnimation)) {
					this.resumeFrom && (this.resumingFrom = this.resumeFrom, this.resumingFrom.resumingFrom = void 0);
					let t = {
						...f(a, "layout"),
						onPlay: o,
						onComplete: s
					};
					(i.shouldReduceMotion || this.options.layoutRoot) && (t.delay = 0, t.type = !1), this.startAnimation(t), this.setAnimationOrigin(e, l, t.path);
				} else t || K(this), this.isLead() && this.options.onExitComplete && this.options.onExitComplete();
				this.targetLayout = r;
			});
		}
		unmount() {
			this.options.layoutId && this.willUpdate(), this.root.nodes.remove(this);
			let e = this.getStack();
			e && e.remove(this), this.parent && this.parent.children.delete(this), this.instance = void 0, this.eventHandlers.clear(), i(this.updateProjection);
		}
		blockUpdate() {
			this.updateManuallyBlocked = !0;
		}
		unblockUpdate() {
			this.updateManuallyBlocked = !1;
		}
		isUpdateBlocked() {
			return this.updateManuallyBlocked || this.updateBlockedByResize;
		}
		isTreeAnimationBlocked() {
			return this.isAnimationBlocked || this.parent && this.parent.isTreeAnimationBlocked() || !1;
		}
		startUpdate() {
			this.isUpdateBlocked() || (this.isUpdating = !0, this.nodes && this.nodes.forEach(Ce), this.animationId++);
		}
		getTransformTemplate() {
			let { visualElement: e } = this.options;
			return e && e.getProps().transformTemplate;
		}
		willUpdate(e = !0) {
			if (this.root.hasTreeAnimated = !0, this.root.isUpdateBlocked()) {
				this.options.onExitComplete && this.options.onExitComplete();
				return;
			}
			if (window.MotionCancelOptimisedAnimation && !this.hasCheckedOptimisedAppear && U(this), !this.root.isUpdating && this.root.startUpdate(), this.isLayoutDirty) return;
			this.isLayoutDirty = !0;
			for (let e = 0; e < this.path.length; e++) {
				let t = this.path[e];
				t.shouldResetTransform = !0, (typeof t.latestValues.x == "string" || typeof t.latestValues.y == "string") && (t.isLayoutDirty = !0), t.updateScroll("snapshot"), t.options.layoutRoot && t.willUpdate(!1);
			}
			let { layoutId: t, layout: n } = this.options;
			if (t === void 0 && !n) return;
			let r = this.getTransformTemplate();
			this.prevTransformTemplateValue = r ? r(this.latestValues, "") : void 0, this.updateSnapshot(), e && this.notifyListeners("willUpdate");
		}
		update() {
			if (this.updateScheduled = !1, this.isUpdateBlocked()) {
				let e = this.updateBlockedByResize;
				this.unblockUpdate(), this.updateBlockedByResize = !1, this.clearAllSnapshots(), e && this.nodes.forEach(ve), this.nodes.forEach(W);
				return;
			}
			if (this.animationId <= this.animationCommitId) {
				this.nodes.forEach(G);
				return;
			}
			this.animationCommitId = this.animationId, this.isUpdating ? (this.isUpdating = !1, this.nodes.forEach(ye), this.nodes.forEach(be), this.nodes.forEach(pe), this.nodes.forEach(me)) : this.nodes.forEach(G), this.clearAllSnapshots();
			let t = c.now();
			o.delta = e(0, 1e3 / 60, t - o.timestamp), o.timestamp = t, o.isProcessing = !0, s.update.process(o), s.preRender.process(o), s.render.process(o), o.isProcessing = !1;
		}
		didUpdate() {
			this.updateScheduled || (this.updateScheduled = !0, m.read(this.scheduleUpdate));
		}
		clearAllSnapshots() {
			this.nodes.forEach(_e), this.sharedNodes.forEach(q);
		}
		scheduleUpdateProjection() {
			this.projectionUpdateScheduled || (this.projectionUpdateScheduled = !0, a.preRender(this.updateProjection, !1, !0));
		}
		scheduleCheckAfterUnmount() {
			a.postRender(() => {
				this.isLayoutDirty ? this.root.didUpdate() : this.root.checkUpdateFailed();
			});
		}
		updateSnapshot() {
			this.snapshot || !this.instance || (this.snapshot = this.measure(), this.snapshot && !D(this.snapshot.measuredBox.x) && !D(this.snapshot.measuredBox.y) && (this.snapshot = void 0));
		}
		updateLayout() {
			if (!this.instance || (this.updateScroll(), !(this.options.alwaysMeasureLayout && this.isLead()) && !this.isLayoutDirty)) return;
			if (this.resumeFrom && !this.resumeFrom.instance) for (let e = 0; e < this.path.length; e++) this.path[e].updateScroll();
			let e = this.layout;
			this.layout = this.measure(!1), this.layoutVersion++, this.layoutCorrected ||= h(), this.isLayoutDirty = !1, this.projectionDelta = void 0, this.notifyListeners("measure", this.layout.layoutBox);
			let { visualElement: t } = this.options;
			t && t.notify("LayoutMeasure", this.layout.layoutBox, e ? e.layoutBox : void 0);
		}
		updateScroll(e = "measure") {
			let t = !!(this.options.layoutScroll && this.instance);
			if (this.scroll && this.scroll.animationId === this.root.animationId && this.scroll.phase === e && (t = !1), t && this.instance) {
				let t = k(this.instance);
				this.scroll = {
					animationId: this.root.animationId,
					phase: e,
					isRoot: t,
					offset: w(this.instance),
					wasRoot: this.scroll ? this.scroll.isRoot : t
				};
			}
		}
		resetTransform() {
			if (!j) return;
			let e = this.isLayoutDirty || this.shouldResetTransform || this.options.alwaysMeasureLayout, t = this.projectionDelta && !P(this.projectionDelta), n = this.getTransformTemplate(), r = n ? n(this.latestValues, "") : void 0, i = r !== this.prevTransformTemplateValue;
			e && this.instance && (t || y(this.latestValues) || i) && (j(this.instance, r), this.shouldResetTransform = !1, this.scheduleRender());
		}
		measure(e = !0) {
			let t = this.measurePageBox(), n = this.removeElementScroll(t);
			return e && (n = this.removeTransform(n)), De(n), {
				animationId: this.root.animationId,
				measuredBox: t,
				layoutBox: n,
				latestValues: {},
				source: this.id
			};
		}
		measurePageBox() {
			let { visualElement: e } = this.options;
			if (!e) return h();
			let t = e.measureViewportBox();
			if (!(this.scroll?.wasRoot || this.path.some(Oe))) {
				let { scroll: e } = this.root;
				e && (x(t.x, e.offset.x), x(t.y, e.offset.y));
			}
			return t;
		}
		removeElementScroll(e) {
			let t = h();
			if (T(t, e), this.scroll?.wasRoot) return t;
			for (let n = 0; n < this.path.length; n++) {
				let r = this.path[n], { scroll: i, options: a } = r;
				r !== this.root && i && a.layoutScroll && (i.wasRoot && T(t, e), x(t.x, i.offset.x), x(t.y, i.offset.y));
			}
			return t;
		}
		applyTransform(e, t = !1, n) {
			let r = n || h();
			T(r, e);
			for (let e = 0; e < this.path.length; e++) {
				let n = this.path[e];
				!t && n.options.layoutScroll && n.scroll && n !== n.root && (x(r.x, -n.scroll.offset.x), x(r.y, -n.scroll.offset.y)), y(n.latestValues) && b(r, n.latestValues, n.layout?.layoutBox);
			}
			return y(this.latestValues) && b(r, this.latestValues, this.layout?.layoutBox), r;
		}
		removeTransform(e) {
			let t = h();
			T(t, e);
			for (let e = 0; e < this.path.length; e++) {
				let n = this.path[e];
				if (!y(n.latestValues)) continue;
				let r;
				n.instance && (v(n.latestValues) && n.updateSnapshot(), r = h(), T(r, n.measurePageBox())), A(t, n.latestValues, n.snapshot?.layoutBox, r);
			}
			return y(this.latestValues) && A(t, this.latestValues), t;
		}
		setTargetDelta(e) {
			this.targetDelta = e, this.root.scheduleUpdateProjection(), this.isProjectionDirty = !0;
		}
		setOptions(e) {
			this.options = {
				...this.options,
				...e,
				crossfade: e.crossfade === void 0 ? !0 : e.crossfade
			};
		}
		clearMeasurements() {
			this.scroll = void 0, this.layout = void 0, this.snapshot = void 0, this.prevTransformTemplateValue = void 0, this.targetDelta = void 0, this.target = void 0, this.isLayoutDirty = !1;
		}
		forceRelativeParentToResolveTarget() {
			this.relativeParent && this.relativeParent.resolvedRelativeTargetAt !== o.timestamp && this.relativeParent.resolveTargetDelta(!0);
		}
		resolveTargetDelta(e = !1) {
			let t = this.getLead();
			this.isProjectionDirty ||= t.isProjectionDirty, this.isTransformDirty ||= t.isTransformDirty, this.isSharedProjectionDirty ||= t.isSharedProjectionDirty;
			let n = !!this.resumingFrom || this !== t;
			if (!(e || n && this.isSharedProjectionDirty || this.isProjectionDirty || this.parent?.isProjectionDirty || this.attemptToResolveRelativeTarget || this.root.updateBlockedByResize)) return;
			let { layout: i, layoutId: a } = this.options;
			if (!this.layout || !(i || a)) return;
			this.resolvedRelativeTargetAt = o.timestamp;
			let s = this.getClosestProjectingParent();
			s && this.linkedParentVersion !== s.layoutVersion && !s.options.layoutRoot && this.removeRelativeTarget(), !this.targetDelta && !this.relativeTarget && (this.options.layoutAnchor !== !1 && s && s.layout ? this.createRelativeTarget(s, this.layout.layoutBox, s.layout.layoutBox) : this.removeRelativeTarget()), !(!this.relativeTarget && !this.targetDelta) && (this.target || (this.target = h(), this.targetWithTransforms = h()), this.relativeTarget && this.relativeTargetOrigin && this.relativeParent && this.relativeParent.target ? (this.forceRelativeParentToResolveTarget(), ie(this.target, this.relativeTarget, this.relativeParent.target, this.options.layoutAnchor || void 0)) : this.targetDelta ? (this.resumingFrom ? this.applyTransform(this.layout.layoutBox, !1, this.target) : T(this.target, this.layout.layoutBox), ne(this.target, this.targetDelta)) : T(this.target, this.layout.layoutBox), this.attemptToResolveRelativeTarget && (this.attemptToResolveRelativeTarget = !1, this.options.layoutAnchor !== !1 && s && !!s.resumingFrom == !!this.resumingFrom && !s.options.layoutScroll && s.target && this.animationProgress !== 1 ? this.createRelativeTarget(s, this.target, s.target) : this.relativeParent = this.relativeTarget = void 0), r.value && R.calculatedTargetDeltas++);
		}
		getClosestProjectingParent() {
			if (!(!this.parent || v(this.parent.latestValues) || _(this.parent.latestValues))) return this.parent.isProjecting() ? this.parent : this.parent.getClosestProjectingParent();
		}
		isProjecting() {
			return !!((this.relativeTarget || this.targetDelta || this.options.layoutRoot) && this.layout);
		}
		createRelativeTarget(e, t, n) {
			this.relativeParent = e, this.linkedParentVersion = e.layoutVersion, this.forceRelativeParentToResolveTarget(), this.relativeTarget = h(), this.relativeTargetOrigin = h(), O(this.relativeTargetOrigin, t, n, this.options.layoutAnchor || void 0), T(this.relativeTarget, this.relativeTargetOrigin);
		}
		removeRelativeTarget() {
			this.relativeParent = this.relativeTarget = void 0;
		}
		calcProjection() {
			let e = this.getLead(), t = !!this.resumingFrom || this !== e, n = !0;
			if ((this.isProjectionDirty || this.parent?.isProjectionDirty) && (n = !1), t && (this.isSharedProjectionDirty || this.isTransformDirty) && (n = !1), this.resolvedRelativeTargetAt === o.timestamp && (n = !1), n) return;
			let { layout: i, layoutId: a } = this.options;
			if (this.isTreeAnimating = !!(this.parent && this.parent.isTreeAnimating || this.currentAnimation || this.pendingAnimation), this.isTreeAnimating || (this.targetDelta = this.relativeTarget = void 0), !this.layout || !(i || a)) return;
			T(this.layoutCorrected, this.layout.layoutBox);
			let s = this.treeScale.x, c = this.treeScale.y;
			re(this.layoutCorrected, this.treeScale, this.path, t), e.layout && !e.target && (this.treeScale.x !== 1 || this.treeScale.y !== 1) && (e.target = e.layout.layoutBox, e.targetWithTransforms = h());
			let { target: l } = e;
			if (!l) {
				this.prevProjectionDelta && (this.createProjectionDeltas(), this.scheduleRender());
				return;
			}
			!this.projectionDelta || !this.prevProjectionDelta ? this.createProjectionDeltas() : (C(this.prevProjectionDelta.x, this.projectionDelta.x), C(this.prevProjectionDelta.y, this.projectionDelta.y)), E(this.projectionDelta, this.layoutCorrected, l, this.latestValues), (this.treeScale.x !== s || this.treeScale.y !== c || !M(this.projectionDelta.x, this.prevProjectionDelta.x) || !M(this.projectionDelta.y, this.prevProjectionDelta.y)) && (this.hasProjected = !0, this.scheduleRender(), this.notifyListeners("projectionUpdate", l)), r.value && R.calculatedProjections++;
		}
		hide() {
			this.isVisible = !1;
		}
		show() {
			this.isVisible = !0;
		}
		scheduleRender(e = !0) {
			if (this.options.visualElement?.scheduleRender(), e) {
				let e = this.getStack();
				e && e.scheduleRender();
			}
			this.resumingFrom && !this.resumingFrom.instance && (this.resumingFrom = void 0);
		}
		createProjectionDeltas() {
			this.prevProjectionDelta = g(), this.projectionDelta = g(), this.projectionDeltaWithTransform = g();
		}
		setAnimationOrigin(e, t = !1, n) {
			let r = this.snapshot, i = r ? r.latestValues : {}, a = { ...this.latestValues }, o = g();
			(!this.relativeParent || !this.relativeParent.options.layoutRoot) && (this.relativeTarget = this.relativeTargetOrigin = void 0), this.attemptToResolveRelativeTarget = !t;
			let s = h(), c = (r ? r.source : void 0) !== (this.layout ? this.layout.source : void 0), l = this.getStack(), d = !l || l.members.length <= 1, f = !!(c && !d && this.options.crossfade === !0 && !this.path.some(Te));
			this.animationProgress = 0;
			let p, m = n?.interpolateProjection(e);
			this.mixTargetDelta = (t) => {
				let n = t / 1e3, r = m?.(n);
				r ? (o.x.translate = r.x, o.x.scale = u(e.x.scale, 1, n), o.x.origin = e.x.origin, o.x.originPoint = e.x.originPoint, o.y.translate = r.y, o.y.scale = u(e.y.scale, 1, n), o.y.origin = e.y.origin, o.y.originPoint = e.y.originPoint) : (J(o.x, e.x, n), J(o.y, e.y, n)), this.setTargetDelta(o), this.relativeTarget && this.relativeTargetOrigin && this.layout && this.relativeParent && this.relativeParent.layout && (O(s, this.layout.layoutBox, this.relativeParent.layout.layoutBox, this.options.layoutAnchor || void 0), we(this.relativeTarget, this.relativeTargetOrigin, s, n), p && ae(this.relativeTarget, p) && (this.isProjectionDirty = !1), p ||= h(), T(p, this.relativeTarget)), c && (this.animationValues = a, se(a, i, this.latestValues, n, f, d)), r && r.rotate !== void 0 && (this.animationValues ||= a, this.animationValues.pathRotation = r.rotate), this.root.scheduleUpdateProjection(), this.scheduleRender(), this.animationProgress = n;
			}, this.mixTargetDelta(this.options.layoutRoot ? 1e3 : 0);
		}
		startAnimation(e) {
			this.notifyListeners("animationStart"), this.currentAnimation?.stop(), this.resumingFrom?.currentAnimation?.stop(), this.pendingAnimation &&= (i(this.pendingAnimation), void 0), this.pendingAnimation = a.update(() => {
				L.hasAnimatedSinceResize = !0, l.layout++, this.motionValue ||= d(0), this.motionValue.jump(0, !1), this.currentAnimation = ce(this.motionValue, [0, 1e3], {
					...e,
					velocity: 0,
					isSync: !0,
					onUpdate: (t) => {
						this.mixTargetDelta(t), e.onUpdate && e.onUpdate(t);
					},
					onStop: () => {
						l.layout--;
					},
					onComplete: () => {
						l.layout--, e.onComplete && e.onComplete(), this.completeAnimation();
					}
				}), this.resumingFrom && (this.resumingFrom.currentAnimation = this.currentAnimation), this.pendingAnimation = void 0;
			});
		}
		completeAnimation() {
			this.resumingFrom && (this.resumingFrom.currentAnimation = void 0, this.resumingFrom.preserveOpacity = void 0);
			let e = this.getStack();
			e && e.exitAnimationComplete(), this.resumingFrom = this.currentAnimation = this.animationValues = void 0, this.notifyListeners("animationComplete");
		}
		finishAnimation() {
			this.currentAnimation && (this.mixTargetDelta && this.mixTargetDelta(B), this.currentAnimation.stop()), this.completeAnimation();
		}
		applyTransformsToTarget() {
			let e = this.getLead(), { targetWithTransforms: t, target: n, layout: r, latestValues: i } = e;
			if (!(!t || !n || !r)) {
				if (this !== e && this.layout && r && $(this.options.animationType, this.layout.layoutBox, r.layoutBox)) {
					n = this.target || h();
					let t = D(this.layout.layoutBox.x);
					n.x.min = e.target.x.min, n.x.max = n.x.min + t;
					let r = D(this.layout.layoutBox.y);
					n.y.min = e.target.y.min, n.y.max = n.y.min + r;
				}
				T(t, n), b(t, i), E(this.projectionDeltaWithTransform, this.layoutCorrected, t, i);
			}
		}
		registerSharedNode(e, t) {
			this.sharedNodes.has(e) || this.sharedNodes.set(e, new de()), this.sharedNodes.get(e).add(t);
			let n = t.options.initialPromotionConfig;
			t.promote({
				transition: n ? n.transition : void 0,
				preserveFollowOpacity: n && n.shouldPreserveFollowOpacity ? n.shouldPreserveFollowOpacity(t) : void 0
			});
		}
		isLead() {
			let e = this.getStack();
			return e ? e.lead === this : !0;
		}
		getLead() {
			let { layoutId: e } = this.options;
			return e && this.getStack()?.lead || this;
		}
		getPrevLead() {
			let { layoutId: e } = this.options;
			return e ? this.getStack()?.prevLead : void 0;
		}
		getStack() {
			let { layoutId: e } = this.options;
			if (e) return this.root.sharedNodes.get(e);
		}
		promote({ needsReset: e, transition: t, preserveFollowOpacity: n } = {}) {
			let r = this.getStack();
			r && r.promote(this, n), e && (this.projectionDelta = void 0, this.needsReset = !0), t && this.setOptions({ transition: t });
		}
		relegate() {
			let e = this.getStack();
			return e ? e.relegate(this) : !1;
		}
		resetSkewAndRotation() {
			let { visualElement: e } = this.options;
			if (!e) return;
			let t = !1, { latestValues: n } = e;
			if ((n.z || n.rotate || n.rotateX || n.rotateY || n.rotateZ || n.skewX || n.skewY) && (t = !0), !t) return;
			let r = {};
			n.z && H("z", e, r, this.animationValues);
			for (let t = 0; t < z.length; t++) H(`rotate${z[t]}`, e, r, this.animationValues), H(`skew${z[t]}`, e, r, this.animationValues);
			e.render();
			for (let t in r) e.setStaticValue(t, r[t]), this.animationValues && (this.animationValues[t] = r[t]);
			e.scheduleRender();
		}
		applyProjectionStyles(e, t) {
			if (!this.instance || this.isSVG) return;
			if (!this.isVisible) {
				e.visibility = "hidden";
				return;
			}
			let n = this.getTransformTemplate();
			if (this.needsReset) {
				this.needsReset = !1, e.visibility = "", e.opacity = "", e.pointerEvents = I(t?.pointerEvents) || "", e.transform = n ? n(this.latestValues, "") : "none";
				return;
			}
			let r = this.getLead();
			if (!this.projectionDelta || !this.layout || !r.target) {
				this.options.layoutId && (e.opacity = this.latestValues.opacity === void 0 ? 1 : this.latestValues.opacity, e.pointerEvents = I(t?.pointerEvents) || ""), this.hasProjected && !y(this.latestValues) && (e.transform = n ? n({}, "") : "none", this.hasProjected = !1);
				return;
			}
			e.visibility = "";
			let i = r.animationValues || r.latestValues;
			this.applyTransformsToTarget();
			let a = oe(this.projectionDeltaWithTransform, this.treeScale, i);
			n && (a = n(i, a)), e.transform = a;
			let { x: o, y: s } = this.projectionDelta;
			e.transformOrigin = `${o.origin * 100}% ${s.origin * 100}% 0`, r.animationValues ? e.opacity = r === this ? i.opacity ?? this.latestValues.opacity ?? 1 : this.preserveOpacity ? this.latestValues.opacity : i.opacityExit : e.opacity = r === this ? i.opacity === void 0 ? "" : i.opacity : i.opacityExit === void 0 ? 0 : i.opacityExit;
			for (let t in S) {
				if (i[t] === void 0) continue;
				let { correct: n, applyTo: o, isCSSVariable: s } = S[t], c = a === "none" ? i[t] : n(i[t], r);
				if (o) {
					let t = o.length;
					for (let n = 0; n < t; n++) e[o[n]] = c;
				} else s ? this.options.visualElement.renderState.vars[t] = c : e[t] = c;
			}
			this.options.layoutId && (e.pointerEvents = r === this ? I(t?.pointerEvents) || "" : "none");
		}
		clearSnapshot() {
			this.resumeFrom = this.snapshot = void 0;
		}
		resetTree() {
			this.root.nodes.forEach((e) => e.currentAnimation?.stop()), this.root.nodes.forEach(W), this.root.sharedNodes.clear();
		}
	};
}
function pe(e) {
	e.updateLayout();
}
function me(e) {
	let t = e.resumeFrom?.snapshot || e.snapshot;
	if (e.isLead() && e.layout && t && e.hasListeners("didUpdate")) {
		let { layoutBox: n, measuredBox: r } = e.layout, { animationType: i } = e.options, a = t.source !== e.layout.source;
		if (i === "size") F((e) => {
			let r = a ? t.measuredBox[e] : t.layoutBox[e], i = D(r);
			r.min = n[e].min, r.max = r.min + i;
		});
		else if (i === "x" || i === "y") {
			let e = i === "x" ? "y" : "x";
			w(a ? t.measuredBox[e] : t.layoutBox[e], n[e]);
		} else $(i, t.layoutBox, n) && F((r) => {
			let i = a ? t.measuredBox[r] : t.layoutBox[r], o = D(n[r]);
			i.max = i.min + o, e.relativeTarget && !e.currentAnimation && (e.isProjectionDirty = !0, e.relativeTarget[r].max = e.relativeTarget[r].min + o);
		});
		let o = g();
		E(o, n, t.layoutBox);
		let s = g();
		a ? E(s, e.applyTransform(r, !0), t.measuredBox) : E(s, n, t.layoutBox);
		let c = !P(o), l = !1;
		if (!e.resumeFrom) {
			let r = e.getClosestProjectingParent();
			if (r && !r.resumeFrom) {
				let { snapshot: i, layout: a } = r;
				if (i && a) {
					let o = e.options.layoutAnchor || void 0, s = h();
					O(s, t.layoutBox, i.layoutBox, o);
					let c = h();
					O(c, n, a.layoutBox, o), N(s, c) || (l = !0), r.options.layoutRoot && (e.relativeTarget = c, e.relativeTargetOrigin = s, e.relativeParent = r);
				}
			}
		}
		e.notifyListeners("didUpdate", {
			layout: n,
			snapshot: t,
			delta: s,
			layoutDelta: o,
			hasLayoutChanged: c,
			hasRelativeLayoutChanged: l
		});
	} else if (e.isLead()) {
		let { onExitComplete: t } = e.options;
		t && t();
	}
	e.options.transition = void 0;
}
function he(e) {
	r.value && R.nodes++, e.parent && (e.isProjecting() || (e.isProjectionDirty = e.parent.isProjectionDirty), e.isSharedProjectionDirty ||= !!(e.isProjectionDirty || e.parent.isProjectionDirty || e.parent.isSharedProjectionDirty), e.isTransformDirty ||= e.parent.isTransformDirty);
}
function ge(e) {
	e.isProjectionDirty = e.isSharedProjectionDirty = e.isTransformDirty = !1;
}
function _e(e) {
	e.clearSnapshot();
}
function W(e) {
	e.clearMeasurements();
}
function ve(e) {
	e.isLayoutDirty = !0, e.updateLayout();
}
function G(e) {
	e.isLayoutDirty = !1;
}
function ye(e) {
	e.isAnimationBlocked && e.layout && !e.isLayoutDirty && (e.snapshot = e.layout, e.isLayoutDirty = !0);
}
function be(e) {
	let { visualElement: t } = e.options;
	t && t.getProps().onBeforeLayoutMeasure && t.notify("BeforeLayoutMeasure"), e.resetTransform();
}
function K(e) {
	e.finishAnimation(), e.targetDelta = e.relativeTarget = e.target = void 0, e.isProjectionDirty = !0;
}
function xe(e) {
	e.resolveTargetDelta();
}
function Se(e) {
	e.calcProjection();
}
function Ce(e) {
	e.resetSkewAndRotation();
}
function q(e) {
	e.removeLeadSnapshot();
}
function J(e, t, n) {
	e.translate = u(t.translate, 0, n), e.scale = u(t.scale, 1, n), e.origin = t.origin, e.originPoint = t.originPoint;
}
function Y(e, t, n, r) {
	e.min = u(t.min, n.min, r), e.max = u(t.max, n.max, r);
}
function we(e, t, n, r) {
	Y(e.x, t.x, n.x, r), Y(e.y, t.y, n.y, r);
}
function Te(e) {
	return e.animationValues && e.animationValues.opacityExit !== void 0;
}
var Ee = {
	duration: .45,
	ease: [
		.4,
		0,
		.1,
		1
	]
}, X = (e) => typeof navigator < "u" && navigator.userAgent && navigator.userAgent.toLowerCase().includes(e), Z = X("applewebkit/") && !X("chrome/") ? Math.round : t;
function Q(e) {
	e.min = Z(e.min), e.max = Z(e.max);
}
function De(e) {
	Q(e.x), Q(e.y);
}
function $(e, t, n) {
	return e === "position" || e === "preserve-aspect" && !k(j(t), j(n), .2);
}
function Oe(e) {
	return e !== e.root && e.scroll?.wasRoot;
}
//#endregion
export { fe as createProjectionNode };
