// motion-dom/dist/es/resize/handle-element.js
import { resolveElements as e } from "../utils/resolve-elements.js";
import { isSVGElement as t } from "../utils/is-svg-element.js";
//#region node_modules/motion-dom/dist/es/resize/handle-element.mjs
var n = /* @__PURE__ */ new WeakMap(), r, i = (e, n, r) => (i, a) => a && a[0] ? a[0][e + "Size"] : t(i) && "getBBox" in i ? i.getBBox()[n] : i[r], a = /*@__PURE__*/ i("inline", "width", "offsetWidth"), o = /*@__PURE__*/ i("block", "height", "offsetHeight");
function s({ target: e, borderBoxSize: t }) {
	n.get(e)?.forEach((n) => {
		n(e, {
			get width() {
				return a(e, t);
			},
			get height() {
				return o(e, t);
			}
		});
	});
}
function c(e) {
	e.forEach(s);
}
function l() {
	typeof ResizeObserver > "u" || (r = new ResizeObserver(c));
}
function u(t, i) {
	r || l();
	let a = e(t);
	return a.forEach((e) => {
		let t = n.get(e);
		t || (t = /* @__PURE__ */ new Set(), n.set(e, t)), t.add(i), r?.observe(e);
	}), () => {
		a.forEach((e) => {
			let t = n.get(e);
			t?.delete(i), t?.size || r?.unobserve(e);
		});
	};
}
//#endregion
export { u as resizeElement };
