// motion-dom/dist/es/animation/keyframes/DOMKeyframesResolver.js
import { containsCSSVariable as e, isCSSVariableToken as t } from "../utils/is-css-variable.js";
import { isNumOrPxType as n, positionalValues as r } from "./utils/unit-conversion.js";
import { KeyframeResolver as i } from "./KeyframesResolver.js";
import { getVariableValue as a } from "../utils/css-variables-conversion.js";
import { positionalKeys as o } from "../../render/utils/keys-position.js";
import { findDimensionValueType as s } from "../../value/types/dimensions.js";
import { isNone as c } from "./utils/is-none.js";
import { makeNoneKeyframesAnimatable as l } from "./utils/make-none-animatable.js";
//#region node_modules/motion-dom/dist/es/animation/keyframes/DOMKeyframesResolver.mjs
var u = class extends i {
	constructor(e, t, n, r, i) {
		super(e, t, n, r, i, !0);
	}
	readKeyframes() {
		let { unresolvedKeyframes: i, element: c, name: l } = this;
		if (!c || !c.current) return;
		super.readKeyframes();
		for (let e = 0; e < i.length; e++) {
			let n = i[e];
			if (typeof n == "string" && (n = n.trim(), t(n))) {
				let t = a(n, c.current);
				t !== void 0 && (i[e] = t), e === i.length - 1 && (this.finalKeyframe = n);
			}
		}
		if (this.resolveNoneKeyframes(), !o.has(l) || i.length !== 2) return;
		let [u, d] = i, f = s(u), p = s(d);
		if (e(u) !== e(d) && r[l]) {
			this.needsMeasurement = !0;
			return;
		}
		if (f !== p) if (n(f) && n(p)) for (let e = 0; e < i.length; e++) {
			let t = i[e];
			typeof t == "string" && (i[e] = parseFloat(t));
		}
		else r[l] && (this.needsMeasurement = !0);
	}
	resolveNoneKeyframes() {
		let { unresolvedKeyframes: e, name: t } = this, n = [];
		for (let t = 0; t < e.length; t++) (e[t] === null || c(e[t])) && n.push(t);
		n.length && l(e, n, t);
	}
	measureInitialState() {
		let { element: e, unresolvedKeyframes: t, name: n } = this;
		if (!e || !e.current) return;
		n === "height" && (this.suspendedScrollY = window.pageYOffset), this.measuredOrigin = r[n](e.measureViewportBox(), window.getComputedStyle(e.current)), t[0] = this.measuredOrigin;
		let i = t[t.length - 1];
		i !== void 0 && e.getValue(n, i).jump(i, !1);
	}
	measureEndState() {
		let { element: e, name: t, unresolvedKeyframes: n } = this;
		if (!e || !e.current) return;
		let i = e.getValue(t);
		i && i.jump(this.measuredOrigin, !1);
		let a = n.length - 1, o = n[a];
		n[a] = r[t](e.measureViewportBox(), window.getComputedStyle(e.current)), o !== null && this.finalKeyframe === void 0 && (this.finalKeyframe = o), this.removedTransforms?.length && this.removedTransforms.forEach(([t, n]) => {
			e.getValue(t).set(n);
		}), this.resolveNoneKeyframes();
	}
};
//#endregion
export { u as DOMKeyframesResolver };
