// motion-dom/dist/es/animation/keyframes/KeyframesResolver.js
import { frame as e } from "../../frameloop/frame.js";
import { fillWildcards as t } from "./utils/fill-wildcards.js";
import { removeNonTranslationalTransform as n } from "./utils/unit-conversion.js";
//#region node_modules/motion-dom/dist/es/animation/keyframes/KeyframesResolver.mjs
var r = /* @__PURE__ */ new Set(), i = !1, a = !1, o = !1;
function s() {
	if (a) {
		let e = Array.from(r).filter((e) => e.needsMeasurement), t = new Set(e.map((e) => e.element)), i = /* @__PURE__ */ new Map();
		t.forEach((e) => {
			let t = n(e);
			t.length && (i.set(e, t), e.render());
		}), e.forEach((e) => e.measureInitialState()), t.forEach((e) => {
			e.render();
			let t = i.get(e);
			t && t.forEach(([t, n]) => {
				e.getValue(t)?.set(n);
			});
		}), e.forEach((e) => e.measureEndState()), e.forEach((e) => {
			e.suspendedScrollY !== void 0 && window.scrollTo(0, e.suspendedScrollY);
		});
	}
	a = !1, i = !1, r.forEach((e) => e.complete(o)), r.clear();
}
function c() {
	r.forEach((e) => {
		e.readKeyframes(), e.needsMeasurement && (a = !0);
	});
}
function l() {
	o = !0, c(), s(), o = !1;
}
var u = class {
	constructor(e, t, n, r, i, a = !1) {
		this.state = "pending", this.isAsync = !1, this.needsMeasurement = !1, this.unresolvedKeyframes = [...e], this.onComplete = t, this.name = n, this.motionValue = r, this.element = i, this.isAsync = a;
	}
	scheduleResolve() {
		this.state = "scheduled", this.isAsync ? (r.add(this), i || (i = !0, e.read(c), e.resolveKeyframes(s))) : (this.readKeyframes(), this.complete());
	}
	readKeyframes() {
		let { unresolvedKeyframes: e, name: n, element: r, motionValue: i } = this;
		if (e[0] === null) {
			let t = i?.get(), a = e[e.length - 1];
			if (t !== void 0) e[0] = t;
			else if (r && n) {
				let t = r.readValue(n, a);
				t != null && (e[0] = t);
			}
			e[0] === void 0 && (e[0] = a), i && t === void 0 && i.set(e[0]);
		}
		t(e);
	}
	setFinalKeyframe() {}
	measureInitialState() {}
	renderEndStyles() {}
	measureEndState() {}
	complete(e = !1) {
		this.state = "complete", this.onComplete(this.unresolvedKeyframes, this.finalKeyframe, e), r.delete(this);
	}
	cancel() {
		this.state === "scheduled" && (r.delete(this), this.state = "pending");
	}
	resume() {
		this.state === "pending" && this.scheduleResolve();
	}
};
//#endregion
export { u as KeyframeResolver, l as flushKeyframeResolvers };
