// framer-motion/dist/es/render/html/use-props.js
"use client";
import { isMotionValue as e } from "../../../../../motion-dom/dist/es/value/utils/is-motion-value.js";
import { buildHTMLStyles as t } from "../../../../../motion-dom/dist/es/render/html/utils/build-styles.js";
import { isForcedMotionValue as n } from "../../../../../motion-dom/dist/es/render/utils/is-forced-motion-value.js";
import { createHtmlRenderState as r } from "./utils/create-render-state.js";
import { useMemo as i } from "react";
//#region node_modules/framer-motion/dist/es/render/html/use-props.mjs
function a(t, r, i) {
	for (let a in r) !e(r[a]) && !n(a, i) && (t[a] = r[a]);
}
function o({ transformTemplate: e }, n) {
	return i(() => {
		let i = r();
		return t(i, n, e), Object.assign({}, i.vars, i.style);
	}, [n]);
}
function s(e, t) {
	let n = e.style || {}, r = {};
	return a(r, n, e), Object.assign(r, o(e, t)), r;
}
function c(e, t) {
	let n = {}, r = s(e, t);
	return e.drag && e.dragListener !== !1 && (n.draggable = !1, r.userSelect = r.WebkitUserSelect = r.WebkitTouchCallout = "none", r.touchAction = e.drag === !0 ? "none" : `pan-${e.drag === "x" ? "y" : "x"}`), e.tabIndex === void 0 && (e.onTap || e.onTapStart || e.whileTap) && (n.tabIndex = 0), n.style = r, n;
}
//#endregion
export { a as copyRawValuesOnly, c as useHTMLProps };
