// motion-dom/dist/es/render/dom/parse-transform.js
//#region node_modules/motion-dom/dist/es/render/dom/parse-transform.mjs
var e = (e) => e * 180 / Math.PI, t = (t) => r(e(Math.atan2(t[1], t[0]))), n = {
	x: 4,
	y: 5,
	translateX: 4,
	translateY: 5,
	scaleX: 0,
	scaleY: 3,
	scale: (e) => (Math.abs(e[0]) + Math.abs(e[3])) / 2,
	rotate: t,
	rotateZ: t,
	skewX: (t) => e(Math.atan(t[1])),
	skewY: (t) => e(Math.atan(t[2])),
	skew: (e) => (Math.abs(e[1]) + Math.abs(e[2])) / 2
}, r = (e) => (e %= 360, e < 0 && (e += 360), e), i = t, a = (e) => Math.sqrt(e[0] * e[0] + e[1] * e[1]), o = (e) => Math.sqrt(e[4] * e[4] + e[5] * e[5]), s = {
	x: 12,
	y: 13,
	z: 14,
	translateX: 12,
	translateY: 13,
	translateZ: 14,
	scaleX: a,
	scaleY: o,
	scale: (e) => (a(e) + o(e)) / 2,
	rotateX: (t) => r(e(Math.atan2(t[6], t[5]))),
	rotateY: (t) => r(e(Math.atan2(-t[2], t[0]))),
	rotateZ: i,
	rotate: i,
	skewX: (t) => e(Math.atan(t[4])),
	skewY: (t) => e(Math.atan(t[1])),
	skew: (e) => (Math.abs(e[1]) + Math.abs(e[4])) / 2
};
function c(e) {
	return +!!e.includes("scale");
}
function l(e, t) {
	if (!e || e === "none") return c(t);
	let r = e.match(/^matrix3d\(([-\d.e\s,]+)\)$/u), i, a;
	if (r) i = s, a = r;
	else {
		let t = e.match(/^matrix\(([-\d.e\s,]+)\)$/u);
		i = n, a = t;
	}
	if (!a) return c(t);
	let o = i[t], l = a[1].split(",").map(d);
	return typeof o == "function" ? o(l) : l[o];
}
var u = (e, t) => {
	let { transform: n = "none" } = getComputedStyle(e);
	return l(n, t);
};
function d(e) {
	return parseFloat(e.trim());
}
//#endregion
export { c as defaultTransformValue, l as parseValueFromTransform, u as readTransformValue };
