// motion-dom/dist/es/render/utils/animation-state.js
import { calcChildStagger as e } from "../../animation/utils/calc-child-stagger.js";
import { resolveVariant as t } from "./resolve-dynamic-variants.js";
import { isKeyframesTarget as n } from "./is-keyframes-target.js";
import { animateVisualElement as r } from "../../animation/interfaces/visual-element.js";
import { isAnimationControls as i } from "./is-animation-controls.js";
import { isVariantLabel as a } from "./is-variant-label.js";
import { variantPriorityOrder as o } from "./variant-props.js";
import { getVariantContext as s } from "./get-variant-context.js";
import { shallowCompare as c } from "./shallow-compare.js";
//#region node_modules/motion-dom/dist/es/render/utils/animation-state.mjs
var l = [...o].reverse(), u = o.length;
function d(e) {
	return (t) => Promise.all(t.map(({ animation: t, options: n }) => r(e, t, n)));
}
function f(r) {
	let o = d(r), f = h(), m = !0, g = !1, _ = (e) => (n, i) => {
		let a = t(r, i, e === "exit" ? r.presenceContext?.custom : void 0);
		if (a) {
			let { transition: e, transitionEnd: t, ...r } = a;
			n = {
				...n,
				...r,
				...t
			};
		}
		return n;
	};
	function v(e) {
		o = e(r);
	}
	function y(d) {
		let { props: h } = r, v = s(r.parent) || {}, y = [], b = /* @__PURE__ */ new Set(), x = {}, S = Infinity;
		for (let o = 0; o < u; o++) {
			let s = l[o], u = f[s], C = h[s] === void 0 ? v[s] : h[s], w = a(C), T = s === d ? u.isActive : null;
			T === !1 && (S = o);
			let E = C === v[s] && C !== h[s] && w;
			if (E && (m || g) && r.manuallyAnimateOnMount && (E = !1), u.protectedKeys = { ...x }, !u.isActive && T === null || !C && !u.prevProp || i(C) || typeof C == "boolean") continue;
			if (s === "exit" && u.isActive && T !== !0) {
				u.prevResolvedValues && (x = {
					...x,
					...u.prevResolvedValues
				});
				continue;
			}
			let D = p(u.prevProp, C), O = D || s === d && u.isActive && !E && w || o > S && w, k = !1, A = Array.isArray(C) ? C : [C], j = A.reduce(_(s), {});
			T === !1 && (j = {});
			let { prevResolvedValues: M = {} } = u, N = {
				...M,
				...j
			}, P = (e) => {
				O = !0, b.has(e) && (k = !0, b.delete(e)), u.needsAnimating[e] = !0;
				let t = r.getValue(e);
				t && (t.liveStyle = !1);
			};
			for (let e in N) {
				let t = j[e], r = M[e];
				if (x.hasOwnProperty(e)) continue;
				let i = !1;
				i = n(t) && n(r) ? !c(t, r) || D : t !== r, i ? t == null ? b.add(e) : P(e) : t !== void 0 && b.has(e) ? P(e) : u.protectedKeys[e] = !0;
			}
			u.prevProp = C, u.prevResolvedValues = j, u.isActive && (x = {
				...x,
				...j
			}), (m || g) && r.blockInitialAnimation && (O = !1);
			let F = E && D;
			O && (!F || k) && y.push(...A.map((n) => {
				let i = { type: s };
				if (typeof n == "string" && (m || g) && !F && r.manuallyAnimateOnMount && r.parent) {
					let { parent: a } = r, o = t(a, n);
					if (a.enteringChildren && o) {
						let { delayChildren: t } = o.transition || {};
						i.delay = e(a.enteringChildren, r, t);
					}
				}
				return {
					animation: n,
					options: i
				};
			}));
		}
		if (b.size) {
			let e = {};
			if (typeof h.initial != "boolean") {
				let n = t(r, Array.isArray(h.initial) ? h.initial[0] : h.initial);
				n && n.transition && (e.transition = n.transition);
			}
			b.forEach((t) => {
				let n = r.getBaseTarget(t), i = r.getValue(t);
				i && (i.liveStyle = !0), e[t] = n ?? null;
			}), y.push({ animation: e });
		}
		let C = !!y.length;
		return m && (h.initial === !1 || h.initial === h.animate) && !r.manuallyAnimateOnMount && (C = !1), m = !1, g = !1, C ? o(y) : Promise.resolve();
	}
	function b(e, t) {
		if (f[e].isActive === t) return Promise.resolve();
		r.variantChildren?.forEach((n) => n.animationState?.setActive(e, t)), f[e].isActive = t;
		let n = y(e);
		for (let e in f) f[e].protectedKeys = {};
		return n;
	}
	return {
		animateChanges: y,
		setActive: b,
		setAnimateFunction: v,
		getState: () => f,
		reset: () => {
			f = h(), g = !0;
		}
	};
}
function p(e, t) {
	return typeof t == "string" ? t !== e : Array.isArray(t) ? !c(t, e) : !1;
}
function m(e = !1) {
	return {
		isActive: e,
		protectedKeys: {},
		needsAnimating: {},
		prevResolvedValues: {}
	};
}
function h() {
	return {
		animate: m(!0),
		whileInView: m(),
		whileHover: m(),
		whileTap: m(),
		whileDrag: m(),
		whileFocus: m(),
		exit: m()
	};
}
//#endregion
export { f as createAnimationState };
