// framer-motion/dist/es/motion/features/layout/MeasureLayout.js
"use client";
import { LayoutGroupContext as e } from "../../../context/LayoutGroupContext.js";
import { frame as t } from "../../../../../../motion-dom/dist/es/frameloop/frame.js";
import { microtask as n } from "../../../../../../motion-dom/dist/es/frameloop/microtask.js";
import { globalProjectionState as r } from "../../../../../../motion-dom/dist/es/projection/node/state.js";
import { usePresence as i } from "../../../components/AnimatePresence/use-presence.js";
import { SwitchLayoutGroupContext as a } from "../../../context/SwitchLayoutGroupContext.js";
import { jsx as o } from "react/jsx-runtime";
import { Component as s, useContext as c } from "react";
//#region node_modules/framer-motion/dist/es/motion/features/layout/MeasureLayout.mjs
var l = !1, u = class extends s {
	componentDidMount() {
		let { visualElement: e, layoutGroup: t, switchLayoutGroup: n, layoutId: i } = this.props, { projection: a } = e;
		a && (t.group && t.group.add(a), n && n.register && i && n.register(a), l && a.root.didUpdate(), a.addEventListener("animationComplete", () => {
			this.safeToRemove();
		}), a.setOptions({
			...a.options,
			layoutDependency: this.props.layoutDependency,
			onExitComplete: () => this.safeToRemove()
		})), r.hasEverUpdated = !0;
	}
	getSnapshotBeforeUpdate(e) {
		let { layoutDependency: n, visualElement: r, drag: i, isPresent: a } = this.props, { projection: o } = r;
		return o ? (o.isPresent = a, e.layoutDependency !== n && o.setOptions({
			...o.options,
			layoutDependency: n
		}), l = !0, i || e.layoutDependency !== n || n === void 0 || e.isPresent !== a ? o.willUpdate() : this.safeToRemove(), e.isPresent !== a && (a ? o.promote() : o.relegate() || t.postRender(() => {
			let e = o.getStack();
			(!e || !e.members.length) && this.safeToRemove();
		})), null) : null;
	}
	componentDidUpdate() {
		let { visualElement: e, layoutAnchor: t } = this.props, { projection: r } = e;
		r && (r.options.layoutAnchor = t, r.root.didUpdate(), n.postRender(() => {
			!r.currentAnimation && r.isLead() && this.safeToRemove();
		}));
	}
	componentWillUnmount() {
		let { visualElement: e, layoutGroup: t, switchLayoutGroup: n } = this.props, { projection: r } = e;
		l = !0, r && (r.scheduleCheckAfterUnmount(), t && t.group && t.group.remove(r), n && n.deregister && n.deregister(r));
	}
	safeToRemove() {
		let { safeToRemove: e } = this.props;
		e && e();
	}
	render() {
		return null;
	}
};
function d(t) {
	let [n, r] = i(), s = c(e);
	return o(u, {
		...t,
		layoutGroup: s,
		switchLayoutGroup: c(a),
		isPresent: n,
		safeToRemove: r
	});
}
//#endregion
export { d as MeasureLayout };
