// motion-dom/dist/es/render/html/HTMLVisualElement.js
import { isCSSVariableName as e } from "../../animation/utils/is-css-variable.js";
import { defaultTransformValue as t, readTransformValue as n } from "../dom/parse-transform.js";
import { transformProps as r } from "../utils/keys-transform.js";
import { DOMVisualElement as i } from "../dom/DOMVisualElement.js";
import { measureViewportBox as a } from "../../projection/utils/measure.js";
import { buildHTMLStyles as o } from "./utils/build-styles.js";
import { renderHTML as s } from "./utils/render.js";
import { scrapeMotionValuesFromProps as c } from "./utils/scrape-motion-values.js";
//#region node_modules/motion-dom/dist/es/render/html/HTMLVisualElement.mjs
function l(e) {
	return window.getComputedStyle(e);
}
var u = class extends i {
	constructor() {
		super(...arguments), this.type = "html", this.renderInstance = s;
	}
	readValueFromInstance(i, a) {
		if (r.has(a)) return this.projection?.isProjecting ? t(a) : n(i, a);
		{
			let t = l(i), n = (e(a) ? t.getPropertyValue(a) : t[a]) || 0;
			return typeof n == "string" ? n.trim() : n;
		}
	}
	measureInstanceViewportBox(e, { transformPagePoint: t }) {
		return a(e, t);
	}
	build(e, t, n) {
		o(e, t, n.transformTemplate);
	}
	scrapeMotionValuesFromProps(e, t, n) {
		return c(e, t, n);
	}
};
//#endregion
export { u as HTMLVisualElement };
