// motion-dom/dist/es/utils/mix/complex.js
import { warning as e } from "../../../../../motion-utils/dist/es/errors.js";
import { pipe as t } from "../../../../../motion-utils/dist/es/pipe.js";
import { isCSSVariableToken as n } from "../../animation/utils/is-css-variable.js";
import { color as r } from "../../value/types/color/index.js";
import { analyseComplexValue as i, complex as a } from "../../value/types/complex/index.js";
import { mixImmediate as o } from "./immediate.js";
import { mixNumber as s } from "./number.js";
import { mixColor as c } from "./color.js";
import { invisibleValues as l, mixVisibility as u } from "./visibility.js";
//#region node_modules/motion-dom/dist/es/utils/mix/complex.mjs
function d(e, t) {
	return (n) => s(e, t, n);
}
function f(e) {
	return typeof e == "number" ? d : typeof e == "string" ? n(e) ? o : r.test(e) ? c : g : Array.isArray(e) ? p : typeof e == "object" ? r.test(e) ? c : m : o;
}
function p(e, t) {
	let n = [...e], r = n.length, i = e.map((e, n) => f(e)(e, t[n]));
	return (e) => {
		for (let t = 0; t < r; t++) n[t] = i[t](e);
		return n;
	};
}
function m(e, t) {
	let n = {
		...e,
		...t
	}, r = {};
	for (let i in n) e[i] !== void 0 && t[i] !== void 0 && (r[i] = f(e[i])(e[i], t[i]));
	return (e) => {
		for (let t in r) n[t] = r[t](e);
		return n;
	};
}
function h(e, t) {
	let n = [], r = {
		color: 0,
		var: 0,
		number: 0
	};
	for (let i = 0; i < t.values.length; i++) {
		let a = t.types[i], o = e.indexes[a][r[a]];
		n[i] = e.values[o] ?? 0, r[a]++;
	}
	return n;
}
var g = (n, r) => {
	let s = a.createTransformer(r), c = i(n), d = i(r);
	return c.indexes.var.length === d.indexes.var.length && c.indexes.color.length === d.indexes.color.length && c.indexes.number.length >= d.indexes.number.length ? l.has(n) && !d.values.length || l.has(r) && !c.values.length ? u(n, r) : t(p(h(c, d), d.values), s) : (e(!0, `Complex values '${n}' and '${r}' too different to mix. Ensure all colors are of the same type, and that each contains the same quantity of number and color values. Falling back to instant transition.`, "complex-values-different"), o(n, r));
};
//#endregion
export { f as getMixer };
