// motion-dom/dist/es/animation/interfaces/motion-value.js
import { MotionGlobalConfig as e } from "../../../../../motion-utils/dist/es/global-config.js";
import { secondsToMilliseconds as t } from "../../../../../motion-utils/dist/es/time-conversion.js";
import { frame as n } from "../../frameloop/frame.js";
import { getFinalKeyframe as r } from "../keyframes/get-final.js";
import { JSAnimation as i } from "../JSAnimation.js";
import { makeAnimationInstant as a } from "../utils/make-animation-instant.js";
import { AsyncMotionValueAnimation as o } from "../AsyncMotionValueAnimation.js";
import { getValueTransition as s } from "../utils/get-value-transition.js";
import { getDefaultTransition as c } from "../utils/default-transitions.js";
import { isTransitionDefined as l } from "../utils/is-transition-defined.js";
//#region node_modules/motion-dom/dist/es/animation/interfaces/motion-value.mjs
var u = (u, d, f, p = {}, m, h) => (g) => {
	let _ = s(p, u) || {}, v = _.delay || p.delay || 0, { elapsed: y = 0 } = p;
	y -= t(v);
	let b = {
		keyframes: Array.isArray(f) ? f : [null, f],
		ease: "easeOut",
		velocity: d.getVelocity(),
		..._,
		delay: -y,
		onUpdate: (e) => {
			d.set(e), _.onUpdate && _.onUpdate(e);
		},
		onComplete: () => {
			g(), _.onComplete && _.onComplete();
		},
		name: u,
		motionValue: d,
		element: h ? void 0 : m
	};
	l(_) || Object.assign(b, c(u, b)), b.duration &&= t(b.duration), b.repeatDelay &&= t(b.repeatDelay), b.from !== void 0 && (b.keyframes[0] = b.from);
	let x = !1;
	if ((b.type === !1 || b.duration === 0 && !b.repeatDelay) && (a(b), b.delay === 0 && (x = !0)), (e.instantAnimations || e.skipAnimations || m?.shouldSkipAnimations || _.skipAnimations) && (x = !0, a(b), b.delay = 0), b.allowFlatten = !_.type && !_.ease, x && !h && d.get() !== void 0) {
		let e = r(b.keyframes, _);
		if (e !== void 0) {
			n.update(() => {
				b.onUpdate(e), b.onComplete();
			});
			return;
		}
	}
	return _.isSync ? new i(b) : new o(b);
};
//#endregion
export { u as animateMotionValue };
