// @base-ui/utils/esm/useMergedRefs.js
import { useRefWithInit as e } from "./useRefWithInit.js";
//#region node_modules/@base-ui/utils/esm/useMergedRefs.js
function t(t, n, a, s) {
	let c = e(r).current;
	return i(c, t, n, a, s) && o(c, [
		t,
		n,
		a,
		s
	]), c.callback;
}
function n(t) {
	let n = e(r).current;
	return a(n, t) && o(n, t), n.callback;
}
function r() {
	return {
		callback: null,
		cleanup: null,
		refs: []
	};
}
function i(e, t, n, r, i) {
	return e.refs[0] !== t || e.refs[1] !== n || e.refs[2] !== r || e.refs[3] !== i;
}
function a(e, t) {
	return e.refs.length !== t.length || e.refs.some((e, n) => e !== t[n]);
}
function o(e, t) {
	if (e.refs = t, t.every((e) => e == null)) {
		e.callback = null;
		return;
	}
	e.callback = (n) => {
		if (e.cleanup &&= (e.cleanup(), null), n != null) {
			let r = Array(t.length).fill(null);
			for (let e = 0; e < t.length; e += 1) {
				let i = t[e];
				if (i != null) switch (typeof i) {
					case "function": {
						let t = i(n);
						typeof t == "function" && (r[e] = t);
						break;
					}
					case "object":
						i.current = n;
						break;
					default:
				}
			}
			e.cleanup = () => {
				for (let e = 0; e < t.length; e += 1) {
					let n = t[e];
					if (n != null) switch (typeof n) {
						case "function": {
							let t = r[e];
							typeof t == "function" ? t() : n(null);
							break;
						}
						case "object":
							n.current = null;
							break;
						default:
					}
				}
			};
		}
	};
}
//#endregion
export { t as useMergedRefs, n as useMergedRefsN };
