// @base-ui/react/esm/merge-props/mergeProps.js
import { mergeObjects as e } from "../../../utils/esm/mergeObjects.js";
//#region node_modules/@base-ui/react/esm/merge-props/mergeProps.js
var t = {};
function n(e, t, n, r, o) {
	if (!n && !r && !o && !e) return i(t);
	let s = i(e);
	return t && (s = a(s, t)), n && (s = a(s, n)), r && (s = a(s, r)), o && (s = a(s, o)), s;
}
function r(e) {
	if (e.length === 0) return t;
	if (e.length === 1) return i(e[0]);
	let n = i(e[0]);
	for (let t = 1; t < e.length; t += 1) n = a(n, e[t]);
	return n;
}
function i(e) {
	return l(e) ? { ...u(e, t) } : o(e);
}
function a(e, t) {
	return l(t) ? u(t, e) : s(e, t);
}
function o(e) {
	let t = { ...e };
	for (let e in t) {
		let n = t[e];
		c(e, n) && (t[e] = f(n));
	}
	return t;
}
function s(t, n) {
	if (!n) return t;
	for (let r in n) {
		let i = n[r];
		switch (r) {
			case "style":
				t[r] = e(t.style, i);
				break;
			case "className":
				t[r] = m(t.className, i);
				break;
			default: c(r, i) ? t[r] = d(t[r], i) : t[r] = i;
		}
	}
	return t;
}
function c(e, t) {
	let n = e.charCodeAt(0), r = e.charCodeAt(1), i = e.charCodeAt(2);
	return n === 111 && r === 110 && i >= 65 && i <= 90 && (typeof t == "function" || t === void 0);
}
function l(e) {
	return typeof e == "function";
}
function u(e, n) {
	return l(e) ? e(n) : e ?? t;
}
function d(e, t) {
	return t ? e ? (...n) => {
		let r = n[0];
		if (h(r)) {
			let i = r;
			p(i);
			let a = t(...n);
			return i.baseUIHandlerPrevented || e?.(...n), a;
		}
		let i = t(...n);
		return e?.(...n), i;
	} : f(t) : e;
}
function f(e) {
	return e && ((...t) => {
		let n = t[0];
		return h(n) && p(n), e(...t);
	});
}
function p(e) {
	return e.preventBaseUIHandler = () => {
		e.baseUIHandlerPrevented = !0;
	}, e;
}
function m(e, t) {
	return t ? e ? t + " " + e : t : e;
}
function h(e) {
	return typeof e == "object" && !!e && "nativeEvent" in e;
}
//#endregion
export { p as makeEventPreventable, m as mergeClassNames, n as mergeProps, r as mergePropsN };
