// motion-dom/dist/es/render/VisualElement.js
import { isNumericalString as e } from "../../../../motion-utils/dist/es/is-numerical-string.js";
import { isZeroValueString as t } from "../../../../motion-utils/dist/es/is-zero-value-string.js";
import { SubscriptionManager as n } from "../../../../motion-utils/dist/es/subscription-manager.js";
import { secondsToMilliseconds as r } from "../../../../motion-utils/dist/es/time-conversion.js";
import { warnOnce as i } from "../../../../motion-utils/dist/es/warn-once.js";
import { cancelFrame as a, frame as o } from "../frameloop/frame.js";
import { time as s } from "../frameloop/sync-time.js";
import { complex as c } from "../value/types/complex/index.js";
import { transformProps as l } from "./utils/keys-transform.js";
import { KeyframeResolver as u } from "../animation/keyframes/KeyframesResolver.js";
import { NativeAnimation as d } from "../animation/NativeAnimation.js";
import { acceleratedValues as f } from "../animation/waapi/utils/accelerated-values.js";
import { motionValue as p } from "../value/index.js";
import { resolveVariantFromProps as m } from "./utils/resolve-variants.js";
import { isMotionValue as h } from "../value/utils/is-motion-value.js";
import { getAnimatableNone as g } from "../value/types/utils/animatable-none.js";
import { microtask as _ } from "../frameloop/microtask.js";
import { findValueType as v } from "../value/types/utils/find.js";
import { createBox as y } from "../projection/geometry/models.js";
import { visualElementStore as b } from "./store.js";
import { isControllingVariants as x, isVariantNode as S } from "./utils/is-controlling-variants.js";
import { updateMotionValuesFromProps as C } from "./utils/motion-values.js";
import { hasReducedMotionListener as w, prefersReducedMotion as T } from "./utils/reduced-motion/state.js";
import { initPrefersReducedMotion as E } from "./utils/reduced-motion/index.js";
//#region node_modules/motion-dom/dist/es/render/VisualElement.mjs
var D = [
	"AnimationStart",
	"AnimationComplete",
	"Update",
	"BeforeLayoutMeasure",
	"LayoutMeasure",
	"LayoutAnimationStart",
	"LayoutAnimationComplete"
], O = {};
function k(e) {
	O = e;
}
function A() {
	return O;
}
var j = class {
	scrapeMotionValuesFromProps(e, t, n) {
		return {};
	}
	constructor({ parent: e, props: t, presenceContext: n, reducedMotionConfig: r, skipAnimations: i, blockInitialAnimation: a, visualState: c }, l = {}) {
		this.current = null, this.children = /* @__PURE__ */ new Set(), this.isVariantNode = !1, this.isControllingVariants = !1, this.shouldReduceMotion = null, this.shouldSkipAnimations = !1, this.values = /* @__PURE__ */ new Map(), this.KeyframeResolver = u, this.features = {}, this.valueSubscriptions = /* @__PURE__ */ new Map(), this.prevMotionValues = {}, this.hasBeenMounted = !1, this.events = {}, this.propEventSubscriptions = {}, this.notifyUpdate = () => this.notify("Update", this.latestValues), this.render = () => {
			this.current && (this.triggerBuild(), this.renderInstance(this.current, this.renderState, this.props.style, this.projection));
		}, this.renderScheduledAt = 0, this.scheduleRender = () => {
			let e = s.now();
			this.renderScheduledAt < e && (this.renderScheduledAt = e, o.render(this.render, !1, !0));
		};
		let { latestValues: d, renderState: f } = c;
		this.latestValues = d, this.baseTarget = { ...d }, this.initialValues = t.initial ? { ...d } : {}, this.renderState = f, this.parent = e, this.props = t, this.presenceContext = n, this.depth = e ? e.depth + 1 : 0, this.reducedMotionConfig = r, this.skipAnimationsConfig = i, this.options = l, this.blockInitialAnimation = !!a, this.isControllingVariants = x(t), this.isVariantNode = S(t), this.isVariantNode && (this.variantChildren = /* @__PURE__ */ new Set()), this.manuallyAnimateOnMount = !!(e && e.current);
		let { willChange: p, ...m } = this.scrapeMotionValuesFromProps(t, {}, this);
		for (let e in m) {
			let t = m[e];
			d[e] !== void 0 && h(t) && t.set(d[e]);
		}
	}
	mount(e) {
		if (this.hasBeenMounted) for (let e in this.initialValues) this.values.get(e)?.jump(this.initialValues[e]), this.latestValues[e] = this.initialValues[e];
		this.current = e, b.set(e, this), this.projection && !this.projection.instance && this.projection.mount(e), this.parent && this.isVariantNode && !this.isControllingVariants && (this.removeFromVariantTree = this.parent.addVariantChild(this)), this.values.forEach((e, t) => this.bindToMotionValue(t, e)), this.reducedMotionConfig === "never" ? this.shouldReduceMotion = !1 : this.reducedMotionConfig === "always" ? this.shouldReduceMotion = !0 : (w.current || E(), this.shouldReduceMotion = T.current), process.env.NODE_ENV !== "production" && i(this.shouldReduceMotion !== !0, "You have Reduced Motion enabled on your device. Animations may not appear as expected.", "reduced-motion-disabled"), this.shouldSkipAnimations = this.skipAnimationsConfig ?? !1, this.parent?.addChild(this), this.update(this.props, this.presenceContext), this.hasBeenMounted = !0;
	}
	unmount() {
		this.projection && this.projection.unmount(), a(this.notifyUpdate), a(this.render), this.valueSubscriptions.forEach((e) => e()), this.valueSubscriptions.clear(), this.removeFromVariantTree && this.removeFromVariantTree(), this.parent?.removeChild(this);
		for (let e in this.events) this.events[e].clear();
		for (let e in this.features) {
			let t = this.features[e];
			t && (t.unmount(), t.isMounted = !1);
		}
		this.current = null;
	}
	addChild(e) {
		this.children.add(e), this.enteringChildren ??= /* @__PURE__ */ new Set(), this.enteringChildren.add(e);
	}
	removeChild(e) {
		this.children.delete(e), this.enteringChildren && this.enteringChildren.delete(e);
	}
	bindToMotionValue(e, t) {
		if (this.valueSubscriptions.has(e) && this.valueSubscriptions.get(e)(), t.accelerate && f.has(e) && this.current instanceof HTMLElement) {
			let { factory: n, keyframes: i, times: a, ease: o, duration: s } = t.accelerate, c = new d({
				element: this.current,
				name: e,
				keyframes: i,
				times: a,
				ease: o,
				duration: r(s)
			}), l = n(c);
			this.valueSubscriptions.set(e, () => {
				l(), c.cancel();
			});
			return;
		}
		let n = l.has(e);
		n && this.onBindTransform && this.onBindTransform();
		let i = t.on("change", (t) => {
			this.latestValues[e] = t, this.props.onUpdate && o.preRender(this.notifyUpdate), n && this.projection && (this.projection.isTransformDirty = !0), this.scheduleRender();
		}), a;
		typeof window < "u" && window.MotionCheckAppearSync && (a = window.MotionCheckAppearSync(this, e, t)), this.valueSubscriptions.set(e, () => {
			i(), a && a();
		});
	}
	sortNodePosition(e) {
		return !this.current || !this.sortInstanceNodePosition || this.type !== e.type ? 0 : this.sortInstanceNodePosition(this.current, e.current);
	}
	updateFeatures() {
		let e = "animation";
		for (e in O) {
			let t = O[e];
			if (!t) continue;
			let { isEnabled: n, Feature: r } = t;
			if (!this.features[e] && r && n(this.props) && (this.features[e] = new r(this)), this.features[e]) {
				let t = this.features[e];
				t.isMounted ? t.update() : (t.mount(), t.isMounted = !0);
			}
		}
	}
	triggerBuild() {
		this.build(this.renderState, this.latestValues, this.props);
	}
	measureViewportBox() {
		return this.current ? this.measureInstanceViewportBox(this.current, this.props) : y();
	}
	getStaticValue(e) {
		return this.latestValues[e];
	}
	setStaticValue(e, t) {
		this.latestValues[e] = t;
	}
	update(e, t) {
		(e.transformTemplate || this.props.transformTemplate) && this.scheduleRender(), this.prevProps = this.props, this.props = e, this.prevPresenceContext = this.presenceContext, this.presenceContext = t;
		for (let t = 0; t < D.length; t++) {
			let n = D[t];
			this.propEventSubscriptions[n] && (this.propEventSubscriptions[n](), delete this.propEventSubscriptions[n]);
			let r = e["on" + n];
			r && (this.propEventSubscriptions[n] = this.on(n, r));
		}
		this.prevMotionValues = C(this, this.scrapeMotionValuesFromProps(e, this.prevProps || {}, this), this.prevMotionValues), this.handleChildMotionValue && this.handleChildMotionValue();
	}
	getProps() {
		return this.props;
	}
	getVariant(e) {
		return this.props.variants ? this.props.variants[e] : void 0;
	}
	getDefaultTransition() {
		return this.props.transition;
	}
	getTransformPagePoint() {
		return this.props.transformPagePoint;
	}
	getClosestVariantNode() {
		return this.isVariantNode ? this : this.parent ? this.parent.getClosestVariantNode() : void 0;
	}
	addVariantChild(e) {
		let t = this.getClosestVariantNode();
		if (t) return t.variantChildren && t.variantChildren.add(e), () => t.variantChildren.delete(e);
	}
	addValue(e, t) {
		let n = this.values.get(e);
		t !== n && (n && this.removeValue(e), this.bindToMotionValue(e, t), this.values.set(e, t), this.latestValues[e] = t.get());
	}
	removeValue(e) {
		this.values.delete(e);
		let t = this.valueSubscriptions.get(e);
		t && (t(), this.valueSubscriptions.delete(e)), delete this.latestValues[e], this.removeValueFromRenderState(e, this.renderState);
	}
	hasValue(e) {
		return this.values.has(e);
	}
	getValue(e, t) {
		if (this.props.values && this.props.values[e]) return this.props.values[e];
		let n = this.values.get(e);
		return n === void 0 && t !== void 0 && (n = p(t === null ? void 0 : t, { owner: this }), this.addValue(e, n)), n;
	}
	readValue(n, r) {
		let i = this.latestValues[n] !== void 0 || !this.current ? this.latestValues[n] : this.getBaseTargetFromProps(this.props, n) ?? this.readValueFromInstance(this.current, n, this.options);
		return i != null && (typeof i == "string" && (e(i) || t(i)) ? i = parseFloat(i) : !v(i) && c.test(r) && (i = g(n, r)), this.setBaseTarget(n, h(i) ? i.get() : i)), h(i) ? i.get() : i;
	}
	setBaseTarget(e, t) {
		this.baseTarget[e] = t;
	}
	getBaseTarget(e) {
		let { initial: t } = this.props, n;
		if (typeof t == "string" || typeof t == "object") {
			let r = m(this.props, t, this.presenceContext?.custom);
			r && (n = r[e]);
		}
		if (t && n !== void 0) return n;
		let r = this.getBaseTargetFromProps(this.props, e);
		return r !== void 0 && !h(r) ? r : this.initialValues[e] !== void 0 && n === void 0 ? void 0 : this.baseTarget[e];
	}
	on(e, t) {
		return this.events[e] || (this.events[e] = new n()), this.events[e].add(t);
	}
	notify(e, ...t) {
		this.events[e] && this.events[e].notify(...t);
	}
	scheduleRenderMicrotask() {
		_.render(this.render);
	}
};
//#endregion
export { j as VisualElement, A as getFeatureDefinitions, k as setFeatureDefinitions };
