// framer-motion/dist/es/gestures/pan/index.js
import { noop as e } from "../../../../../motion-utils/dist/es/noop.js";
import { frame as t } from "../../../../../motion-dom/dist/es/frameloop/frame.js";
import { Feature as n } from "../../../../../motion-dom/dist/es/render/Feature.js";
import { addPointerEvent as r } from "../../events/add-pointer-event.js";
import { getContextWindow as i } from "../../utils/get-context-window.js";
import { PanSession as a } from "./PanSession.js";
//#region node_modules/framer-motion/dist/es/gestures/pan/index.mjs
var o = (e) => (n, r) => {
	e && t.update(() => e(n, r), !1, !0);
}, s = class extends n {
	constructor() {
		super(...arguments), this.removePointerDownListener = e;
	}
	onPointerDown(e) {
		this.session = new a(e, this.createPanHandlers(), {
			transformPagePoint: this.node.getTransformPagePoint(),
			contextWindow: i(this.node)
		});
	}
	createPanHandlers() {
		let { onPanSessionStart: e, onPanStart: n, onPan: r, onPanEnd: i } = this.node.getProps();
		return {
			onSessionStart: o(e),
			onStart: o(n),
			onMove: o(r),
			onEnd: (e, n) => {
				delete this.session, i && t.postRender(() => i(e, n));
			}
		};
	}
	mount() {
		this.removePointerDownListener = r(this.node.current, "pointerdown", (e) => this.onPointerDown(e));
	}
	update() {
		this.session && this.session.updateHandlers(this.createPanHandlers());
	}
	unmount() {
		this.removePointerDownListener(), this.session && this.session.end();
	}
};
//#endregion
export { s as PanGesture };
