// framer-motion/dist/es/gestures/pan/PanSession.js
import { pipe as e } from "../../../../../motion-utils/dist/es/pipe.js";
import { millisecondsToSeconds as t, secondsToMilliseconds as n } from "../../../../../motion-utils/dist/es/time-conversion.js";
import { cancelFrame as r, frame as i, frameData as a } from "../../../../../motion-dom/dist/es/frameloop/frame.js";
import { isPrimaryPointer as o } from "../../../../../motion-dom/dist/es/gestures/utils/is-primary-pointer.js";
import { extractEventInfo as s } from "../../events/event-info.js";
import { addPointerEvent as c } from "../../events/add-pointer-event.js";
import { distance2D as l } from "../../utils/distance.js";
//#region node_modules/framer-motion/dist/es/gestures/pan/PanSession.mjs
var u = /*#__PURE__*/ new Set(["auto", "scroll"]), d = class {
	constructor(t, n, { transformPagePoint: r, contextWindow: u = window, dragSnapToOrigin: d = !1, distanceThreshold: p = 3, element: h } = {}) {
		if (this.startEvent = null, this.lastMoveEvent = null, this.lastMoveEventInfo = null, this.lastRawMoveEventInfo = null, this.handlers = {}, this.contextWindow = window, this.scrollPositions = /* @__PURE__ */ new Map(), this.removeScrollListeners = null, this.onElementScroll = (e) => {
			this.handleScroll(e.target);
		}, this.onWindowScroll = () => {
			this.handleScroll(window);
		}, this.updatePoint = () => {
			if (!(this.lastMoveEvent && this.lastMoveEventInfo)) return;
			this.lastRawMoveEventInfo && (this.lastMoveEventInfo = f(this.lastRawMoveEventInfo, this.transformPagePoint));
			let e = m(this.lastMoveEventInfo, this.history), t = this.startEvent !== null, n = l(e.offset, {
				x: 0,
				y: 0
			}) >= this.distanceThreshold;
			if (!t && !n) return;
			let { point: r } = e, { timestamp: i } = a;
			this.history.push({
				...r,
				timestamp: i
			});
			let { onStart: o, onMove: s } = this.handlers;
			t || (o && o(this.lastMoveEvent, e), this.startEvent = this.lastMoveEvent), s && s(this.lastMoveEvent, e);
		}, this.handlePointerMove = (e, t) => {
			this.lastMoveEvent = e, this.lastRawMoveEventInfo = t, this.lastMoveEventInfo = f(t, this.transformPagePoint), i.update(this.updatePoint, !0);
		}, this.handlePointerUp = (e, t) => {
			this.end();
			let { onEnd: n, onSessionEnd: r, resumeAnimation: i } = this.handlers;
			if ((this.dragSnapToOrigin || !this.startEvent) && i && i(), !(this.lastMoveEvent && this.lastMoveEventInfo)) return;
			let a = m(e.type === "pointercancel" ? this.lastMoveEventInfo : f(t, this.transformPagePoint), this.history);
			this.startEvent && n && n(e, a), r && r(e, a);
		}, !o(t)) return;
		this.dragSnapToOrigin = d, this.handlers = n, this.transformPagePoint = r, this.distanceThreshold = p, this.contextWindow = u || window;
		let g = f(s(t), this.transformPagePoint), { point: _ } = g, { timestamp: v } = a;
		this.history = [{
			..._,
			timestamp: v
		}];
		let { onSessionStart: y } = n;
		y && y(t, m(g, this.history)), this.removeListeners = e(c(this.contextWindow, "pointermove", this.handlePointerMove), c(this.contextWindow, "pointerup", this.handlePointerUp), c(this.contextWindow, "pointercancel", this.handlePointerUp)), h && this.startScrollTracking(h);
	}
	startScrollTracking(e) {
		let t = e.parentElement;
		for (; t;) {
			let e = getComputedStyle(t);
			(u.has(e.overflowX) || u.has(e.overflowY)) && this.scrollPositions.set(t, {
				x: t.scrollLeft,
				y: t.scrollTop
			}), t = t.parentElement;
		}
		this.scrollPositions.set(window, {
			x: window.scrollX,
			y: window.scrollY
		}), window.addEventListener("scroll", this.onElementScroll, { capture: !0 }), window.addEventListener("scroll", this.onWindowScroll), this.removeScrollListeners = () => {
			window.removeEventListener("scroll", this.onElementScroll, { capture: !0 }), window.removeEventListener("scroll", this.onWindowScroll);
		};
	}
	handleScroll(e) {
		let t = this.scrollPositions.get(e);
		if (!t) return;
		let n = e === window, r = n ? {
			x: window.scrollX,
			y: window.scrollY
		} : {
			x: e.scrollLeft,
			y: e.scrollTop
		}, a = {
			x: r.x - t.x,
			y: r.y - t.y
		};
		a.x === 0 && a.y === 0 || (n ? this.lastMoveEventInfo && (this.lastMoveEventInfo.point.x += a.x, this.lastMoveEventInfo.point.y += a.y) : this.history.length > 0 && (this.history[0].x -= a.x, this.history[0].y -= a.y), this.scrollPositions.set(e, r), i.update(this.updatePoint, !0));
	}
	updateHandlers(e) {
		this.handlers = e;
	}
	end() {
		this.removeListeners && this.removeListeners(), this.removeScrollListeners && this.removeScrollListeners(), this.scrollPositions.clear(), r(this.updatePoint);
	}
};
function f(e, t) {
	return t ? { point: t(e.point) } : e;
}
function p(e, t) {
	return {
		x: e.x - t.x,
		y: e.y - t.y
	};
}
function m({ point: e }, t) {
	return {
		point: e,
		delta: p(e, g(t)),
		offset: p(e, h(t)),
		velocity: _(t, .1)
	};
}
function h(e) {
	return e[0];
}
function g(e) {
	return e[e.length - 1];
}
function _(e, r) {
	if (e.length < 2) return {
		x: 0,
		y: 0
	};
	let i = e.length - 1, a = null, o = g(e);
	for (; i >= 0 && (a = e[i], !(o.timestamp - a.timestamp > n(r)));) i--;
	if (!a) return {
		x: 0,
		y: 0
	};
	a === e[0] && e.length > 2 && o.timestamp - a.timestamp > n(r) * 2 && (a = e[1]);
	let s = t(o.timestamp - a.timestamp);
	if (s === 0) return {
		x: 0,
		y: 0
	};
	let c = {
		x: (o.x - a.x) / s,
		y: (o.y - a.y) / s
	};
	return c.x === Infinity && (c.x = 0), c.y === Infinity && (c.y = 0), c;
}
//#endregion
export { d as PanSession };
