// framer-motion/dist/es/gestures/press.js
import { frame as e } from "../../../../motion-dom/dist/es/frameloop/frame.js";
import { press as t } from "../../../../motion-dom/dist/es/gestures/press/index.js";
import { Feature as n } from "../../../../motion-dom/dist/es/render/Feature.js";
import { extractEventInfo as r } from "../events/event-info.js";
//#region node_modules/framer-motion/dist/es/gestures/press.mjs
function i(t, n, i) {
	let { props: a } = t;
	if (t.current instanceof HTMLButtonElement && t.current.disabled) return;
	t.animationState && a.whileTap && t.animationState.setActive("whileTap", i === "Start");
	let o = a["onTap" + (i === "End" ? "" : i)];
	o && e.postRender(() => o(n, r(n)));
}
var a = class extends n {
	mount() {
		let { current: e } = this.node;
		if (!e) return;
		let { globalTapTarget: n, propagate: r } = this.node.props;
		this.unmount = t(e, (e, t) => (i(this.node, t, "Start"), (e, { success: t }) => i(this.node, e, t ? "End" : "Cancel")), {
			useGlobalTarget: n,
			stopPropagation: r?.tap === !1
		});
	}
	unmount() {}
};
//#endregion
export { a as PressGesture };
