// framer-motion/dist/es/gestures/drag/index.js
import { noop as e } from "../../../../../motion-utils/dist/es/noop.js";
import { Feature as t } from "../../../../../motion-dom/dist/es/render/Feature.js";
import { VisualElementDragControls as n } from "./VisualElementDragControls.js";
//#region node_modules/framer-motion/dist/es/gestures/drag/index.mjs
var r = class extends t {
	constructor(t) {
		super(t), this.removeGroupControls = e, this.removeListeners = e, this.controls = new n(t);
	}
	mount() {
		let { dragControls: t } = this.node.getProps();
		t && (this.removeGroupControls = t.subscribe(this.controls)), this.removeListeners = this.controls.addListeners() || e;
	}
	update() {
		let { dragControls: e } = this.node.getProps(), { dragControls: t } = this.node.prevProps || {};
		e !== t && (this.removeGroupControls(), e && (this.removeGroupControls = e.subscribe(this.controls)));
	}
	unmount() {
		this.removeGroupControls(), this.removeListeners(), this.controls.isDragging || this.controls.endPanSession();
	}
};
//#endregion
export { r as DragGesture };
