// framer-motion/dist/es/gestures/drag/VisualElementDragControls.js
import { invariant as e } from "../../../../../motion-utils/dist/es/errors.js";
import { frame as t } from "../../../../../motion-dom/dist/es/frameloop/frame.js";
import { percent as n } from "../../../../../motion-dom/dist/es/value/types/numbers/units.js";
import { mixNumber as r } from "../../../../../motion-dom/dist/es/utils/mix/number.js";
import { animateMotionValue as i } from "../../../../../motion-dom/dist/es/animation/interfaces/motion-value.js";
import { addValueToWillChange as a } from "../../../../../motion-dom/dist/es/value/will-change/add-will-change.js";
import { setDragLock as o } from "../../../../../motion-dom/dist/es/gestures/drag/state/set-active.js";
import { isElementTextInput as s } from "../../../../../motion-dom/dist/es/gestures/press/utils/is-keyboard-accessible.js";
import { resize as c } from "../../../../../motion-dom/dist/es/resize/index.js";
import { createBox as l } from "../../../../../motion-dom/dist/es/projection/geometry/models.js";
import { convertBoundingBoxToBox as u, convertBoxToBoundingBox as d } from "../../../../../motion-dom/dist/es/projection/geometry/conversion.js";
import { measurePageBox as f } from "../../../../../motion-dom/dist/es/projection/utils/measure.js";
import { calcLength as p } from "../../../../../motion-dom/dist/es/projection/geometry/delta-calc.js";
import { eachAxis as m } from "../../../../../motion-dom/dist/es/projection/utils/each-axis.js";
import { addDomEvent as h } from "../../../../../motion-dom/dist/es/events/add-dom-event.js";
import { isRefObject as g } from "../../utils/is-ref-object.js";
import { extractEventInfo as _ } from "../../events/event-info.js";
import { addPointerEvent as v } from "../../events/add-pointer-event.js";
import { getContextWindow as y } from "../../utils/get-context-window.js";
import { PanSession as b } from "../pan/PanSession.js";
import { applyConstraints as x, calcOrigin as S, calcRelativeConstraints as C, calcViewportConstraints as w, defaultElastic as T, rebaseAxisConstraints as E, resolveDragElastic as D } from "./utils/constraints.js";
//#region node_modules/framer-motion/dist/es/gestures/drag/VisualElementDragControls.mjs
var O = /* @__PURE__ */ new WeakMap(), k = class {
	constructor(e) {
		this.openDragLock = null, this.isDragging = !1, this.currentDirection = null, this.originPoint = {
			x: 0,
			y: 0
		}, this.constraints = !1, this.hasMutatedConstraints = !1, this.elastic = l(), this.latestPointerEvent = null, this.latestPanInfo = null, this.visualElement = e;
	}
	start(e, { snapToCursor: r = !1, distanceThreshold: i } = {}) {
		let { presenceContext: s } = this.visualElement;
		if (s && s.isPresent === !1) return;
		let c = (e) => {
			r && this.snapToCursor(_(e).point), this.stopAnimation();
		}, l = (e, r) => {
			let { drag: i, dragPropagation: s, onDragStart: c } = this.getProps();
			if (i && !s && (this.openDragLock && this.openDragLock(), this.openDragLock = o(i), !this.openDragLock)) return;
			this.latestPointerEvent = e, this.latestPanInfo = r, this.isDragging = !0, this.currentDirection = null, this.resolveConstraints(), this.visualElement.projection && (this.visualElement.projection.isAnimationBlocked = !0, this.visualElement.projection.target = void 0), m((e) => {
				let t = this.getAxisMotionValue(e).get() || 0;
				if (n.test(t)) {
					let { projection: n } = this.visualElement;
					if (n && n.layout) {
						let r = n.layout.layoutBox[e];
						r && (t = p(r) * (parseFloat(t) / 100));
					}
				}
				this.originPoint[e] = t;
			}), c && t.update(() => c(e, r), !1, !0), a(this.visualElement, "transform");
			let { animationState: l } = this.visualElement;
			l && l.setActive("whileDrag", !0);
		}, u = (e, n) => {
			this.latestPointerEvent = e, this.latestPanInfo = n;
			let { dragPropagation: r, dragDirectionLock: i, onDirectionLock: a, onDrag: o } = this.getProps();
			if (!r && !this.openDragLock) return;
			let { offset: s } = n;
			if (i && this.currentDirection === null) {
				this.currentDirection = N(s), this.currentDirection !== null && a && a(this.currentDirection);
				return;
			}
			this.updateAxis("x", n.point, s), this.updateAxis("y", n.point, s), this.visualElement.render(), o && t.update(() => o(e, n), !1, !0);
		}, d = (e, t) => {
			this.latestPointerEvent = e, this.latestPanInfo = t, this.stop(e, t), this.latestPointerEvent = null, this.latestPanInfo = null;
		}, f = () => {
			let { dragSnapToOrigin: e } = this.getProps();
			(e || this.constraints) && this.startAnimation({
				x: 0,
				y: 0
			});
		}, { dragSnapToOrigin: h } = this.getProps();
		this.panSession = new b(e, {
			onSessionStart: c,
			onStart: l,
			onMove: u,
			onSessionEnd: d,
			resumeAnimation: f
		}, {
			transformPagePoint: this.visualElement.getTransformPagePoint(),
			dragSnapToOrigin: h,
			distanceThreshold: i,
			contextWindow: y(this.visualElement),
			element: this.visualElement.current
		});
	}
	stop(e, n) {
		let r = e || this.latestPointerEvent, i = n || this.latestPanInfo, a = this.isDragging;
		if (this.cancel(), !a || !i || !r) return;
		let { velocity: o } = i;
		this.startAnimation(o);
		let { onDragEnd: s } = this.getProps();
		s && t.postRender(() => s(r, i));
	}
	cancel() {
		this.isDragging = !1;
		let { projection: e, animationState: t } = this.visualElement;
		e && (e.isAnimationBlocked = !1), this.endPanSession();
		let { dragPropagation: n } = this.getProps();
		!n && this.openDragLock && (this.openDragLock(), this.openDragLock = null), t && t.setActive("whileDrag", !1);
	}
	endPanSession() {
		this.panSession && this.panSession.end(), this.panSession = void 0;
	}
	updateAxis(e, t, n) {
		let { drag: r } = this.getProps();
		if (!n || !M(e, r, this.currentDirection)) return;
		let i = this.getAxisMotionValue(e), a = this.originPoint[e] + n[e];
		this.constraints && this.constraints[e] && (a = x(a, this.constraints[e], this.elastic[e])), i.set(a);
	}
	resolveConstraints() {
		let { dragConstraints: e, dragElastic: t } = this.getProps(), n = this.visualElement.projection && !this.visualElement.projection.layout ? this.visualElement.projection.measure(!1) : this.visualElement.projection?.layout, r = this.constraints;
		e && g(e) ? this.constraints ||= this.resolveRefConstraints() : e && n ? this.constraints = C(n.layoutBox, e) : this.constraints = !1, this.elastic = D(t), r !== this.constraints && !g(e) && n && this.constraints && !this.hasMutatedConstraints && m((e) => {
			this.constraints !== !1 && this.getAxisMotionValue(e) && (this.constraints[e] = E(n.layoutBox[e], this.constraints[e]));
		});
	}
	resolveRefConstraints() {
		let { dragConstraints: t, onMeasureDragConstraints: n } = this.getProps();
		if (!t || !g(t)) return !1;
		let r = t.current;
		e(r !== null, "If `dragConstraints` is set as a React ref, that ref must be passed to another component's `ref` prop.", "drag-constraints-ref");
		let { projection: i } = this.visualElement;
		if (!i || !i.layout) return !1;
		i.root && (i.root.scroll = void 0, i.root.updateScroll());
		let a = f(r, i.root, this.visualElement.getTransformPagePoint()), o = w(i.layout.layoutBox, a);
		if (n) {
			let e = n(d(o));
			this.hasMutatedConstraints = !!e, e && (o = u(e));
		}
		return o;
	}
	startAnimation(e) {
		let { drag: t, dragMomentum: n, dragElastic: r, dragTransition: i, dragSnapToOrigin: a, onDragTransitionEnd: o } = this.getProps(), s = this.constraints || {}, c = m((o) => {
			if (!M(o, t, this.currentDirection)) return;
			let c = s && s[o] || {};
			(a === !0 || a === o) && (c = {
				min: 0,
				max: 0
			});
			let l = r ? 200 : 1e6, u = r ? 40 : 1e7, d = {
				type: "inertia",
				velocity: n ? e[o] : 0,
				bounceStiffness: l,
				bounceDamping: u,
				timeConstant: 750,
				restDelta: 1,
				restSpeed: 10,
				...i,
				...c
			};
			return this.startAxisValueAnimation(o, d);
		});
		return Promise.all(c).then(o);
	}
	startAxisValueAnimation(e, t) {
		let n = this.getAxisMotionValue(e);
		return a(this.visualElement, e), n.start(i(e, n, 0, t, this.visualElement, !1));
	}
	stopAnimation() {
		m((e) => this.getAxisMotionValue(e).stop());
	}
	getAxisMotionValue(e) {
		let t = `_drag${e.toUpperCase()}`;
		return this.visualElement.getProps()[t] || this.visualElement.getValue(e, this.visualElement.latestValues[e] ?? 0);
	}
	snapToCursor(e) {
		m((t) => {
			let { drag: n } = this.getProps();
			if (!M(t, n, this.currentDirection)) return;
			let { projection: i } = this.visualElement, a = this.getAxisMotionValue(t);
			if (i && i.layout) {
				let { min: n, max: o } = i.layout.layoutBox[t], s = a.get() || 0;
				a.set(e[t] - r(n, o, .5) + s);
			}
		});
	}
	scalePositionWithinConstraints() {
		if (!this.visualElement.current) return;
		let { drag: e, dragConstraints: t } = this.getProps(), { projection: n } = this.visualElement;
		if (!g(t) || !n || !this.constraints) return;
		this.stopAnimation();
		let i = {
			x: 0,
			y: 0
		};
		m((e) => {
			let t = this.getAxisMotionValue(e);
			if (t && this.constraints !== !1) {
				let n = t.get();
				i[e] = S({
					min: n,
					max: n
				}, this.constraints[e]);
			}
		});
		let { transformTemplate: a } = this.visualElement.getProps();
		this.visualElement.current.style.transform = a ? a({}, "") : "none", n.root && n.root.updateScroll(), n.updateLayout(), this.constraints = !1, this.resolveConstraints(), m((t) => {
			if (!M(t, e, null)) return;
			let n = this.getAxisMotionValue(t), { min: a, max: o } = this.constraints[t];
			n.set(r(a, o, i[t]));
		}), this.visualElement.render();
	}
	addListeners() {
		if (!this.visualElement.current) return;
		O.set(this.visualElement, this);
		let e = this.visualElement.current, n = v(e, "pointerdown", (t) => {
			let { drag: n, dragListener: r = !0 } = this.getProps(), i = t.target, a = i !== e && s(i);
			n && r && !a && this.start(t);
		}), r, i = () => {
			let { dragConstraints: t } = this.getProps();
			g(t) && t.current && (this.constraints = this.resolveRefConstraints(), r ||= j(e, t.current, () => this.scalePositionWithinConstraints()));
		}, { projection: a } = this.visualElement, o = a.addEventListener("measure", i);
		a && !a.layout && (a.root && a.root.updateScroll(), a.updateLayout()), t.read(i);
		let c = h(window, "resize", () => this.scalePositionWithinConstraints()), l = a.addEventListener("didUpdate", (({ delta: e, hasLayoutChanged: t }) => {
			this.isDragging && t && (m((t) => {
				let n = this.getAxisMotionValue(t);
				n && (this.originPoint[t] += e[t].translate, n.set(n.get() + e[t].translate));
			}), this.visualElement.render());
		}));
		return () => {
			c(), n(), o(), l && l(), r && r();
		};
	}
	getProps() {
		let e = this.visualElement.getProps(), { drag: t = !1, dragDirectionLock: n = !1, dragPropagation: r = !1, dragConstraints: i = !1, dragElastic: a = T, dragMomentum: o = !0 } = e;
		return {
			...e,
			drag: t,
			dragDirectionLock: n,
			dragPropagation: r,
			dragConstraints: i,
			dragElastic: a,
			dragMomentum: o
		};
	}
};
function A(e) {
	let t = !0;
	return () => {
		if (t) {
			t = !1;
			return;
		}
		e();
	};
}
function j(e, t, n) {
	let r = c(e, A(n)), i = c(t, A(n));
	return () => {
		r(), i();
	};
}
function M(e, t, n) {
	return (t === !0 || t === e) && (n === null || n === e);
}
function N(e, t = 10) {
	let n = null;
	return Math.abs(e.y) > t ? n = "y" : Math.abs(e.x) > t && (n = "x"), n;
}
//#endregion
export { k as VisualElementDragControls };
