// motion-utils/dist/es/easing/utils/map.js
import { invariant as e } from "../../errors.js";
import { noop as t } from "../../noop.js";
import { cubicBezier as n } from "../cubic-bezier.js";
import { backIn as r, backInOut as i, backOut as a } from "../back.js";
import { anticipate as o } from "../anticipate.js";
import { circIn as s, circInOut as c, circOut as l } from "../circ.js";
import { easeIn as u, easeInOut as d, easeOut as f } from "../ease.js";
import { isBezierDefinition as p } from "./is-bezier-definition.js";
//#region node_modules/motion-utils/dist/es/easing/utils/map.mjs
var m = {
	linear: t,
	easeIn: u,
	easeInOut: d,
	easeOut: f,
	circIn: s,
	circInOut: c,
	circOut: l,
	backIn: r,
	backInOut: i,
	backOut: a,
	anticipate: o
}, h = (e) => typeof e == "string", g = (t) => {
	if (p(t)) {
		e(t.length === 4, "Cubic bezier arrays must contain four numerical values.", "cubic-bezier-length");
		let [r, i, a, o] = t;
		return n(r, i, a, o);
	} else if (h(t)) return e(m[t] !== void 0, `Invalid easing type '${t}'`, "invalid-easing-type"), m[t];
	return t;
};
//#endregion
export { g as easingDefinitionToFunction };
