// motion-dom/dist/es/animation/keyframes/utils/unit-conversion.js
import { number as e } from "../../../value/types/numbers/index.js";
import { px as t } from "../../../value/types/numbers/units.js";
import { parseValueFromTransform as n } from "../../../render/dom/parse-transform.js";
import { transformPropOrder as r } from "../../../render/utils/keys-transform.js";
//#region node_modules/motion-dom/dist/es/animation/keyframes/utils/unit-conversion.mjs
var i = (n) => n === e || n === t, a = new Set([
	"x",
	"y",
	"z"
]), o = r.filter((e) => !a.has(e));
function s(e) {
	let t = [];
	return o.forEach((n) => {
		let r = e.getValue(n);
		r !== void 0 && (t.push([n, r.get()]), r.set(+!!n.startsWith("scale")));
	}), t;
}
var c = {
	width: ({ x: e }, { paddingLeft: t = "0", paddingRight: n = "0", boxSizing: r }) => {
		let i = e.max - e.min;
		return r === "border-box" ? i : i - parseFloat(t) - parseFloat(n);
	},
	height: ({ y: e }, { paddingTop: t = "0", paddingBottom: n = "0", boxSizing: r }) => {
		let i = e.max - e.min;
		return r === "border-box" ? i : i - parseFloat(t) - parseFloat(n);
	},
	top: (e, { top: t }) => parseFloat(t),
	left: (e, { left: t }) => parseFloat(t),
	bottom: ({ y: e }, { top: t }) => parseFloat(t) + (e.max - e.min),
	right: ({ x: e }, { left: t }) => parseFloat(t) + (e.max - e.min),
	x: (e, { transform: t }) => n(t, "x"),
	y: (e, { transform: t }) => n(t, "y")
};
c.translateX = c.x, c.translateY = c.y;
//#endregion
export { i as isNumOrPxType, c as positionalValues, s as removeNonTranslationalTransform };
