// framer-motion/dist/es/motion/utils/use-visual-element.js
"use client";
import { useIsomorphicLayoutEffect as e } from "../../utils/use-isomorphic-effect.js";
import { PresenceContext as t } from "../../context/PresenceContext.js";
import { optimizedAppearDataAttribute as n } from "../../../../../motion-dom/dist/es/animation/optimized-appear/data-id.js";
import { MotionConfigContext as r } from "../../context/MotionConfigContext.js";
import { LazyContext as i } from "../../context/LazyContext.js";
import { MotionContext as a } from "../../context/MotionContext/index.js";
import { SwitchLayoutGroupContext as o } from "../../context/SwitchLayoutGroupContext.js";
import { isRefObject as s } from "../../utils/is-ref-object.js";
import { useContext as c, useEffect as l, useInsertionEffect as u, useRef as d } from "react";
//#region node_modules/framer-motion/dist/es/motion/utils/use-visual-element.mjs
function f(s, f, m, h, g, _) {
	let { visualElement: v } = c(a), y = c(i), b = c(t), x = c(r), S = x.reducedMotion, C = x.skipAnimations, w = d(null), T = d(!1);
	h ||= y.renderer, !w.current && h && (w.current = h(s, {
		visualState: f,
		parent: v,
		props: m,
		presenceContext: b,
		blockInitialAnimation: b ? b.initial === !1 : !1,
		reducedMotionConfig: S,
		skipAnimations: C,
		isSVG: _
	}), T.current && w.current && (w.current.manuallyAnimateOnMount = !0));
	let E = w.current, D = c(o);
	E && !E.projection && g && (E.type === "html" || E.type === "svg") && p(w.current, m, g, D);
	let O = d(!1);
	u(() => {
		E && O.current && E.update(m, b);
	});
	let k = m[n], A = d(!!k && typeof window < "u" && !window.MotionHandoffIsComplete?.(k) && window.MotionHasOptimisedAnimation?.(k));
	return e(() => {
		T.current = !0, E && (O.current = !0, window.MotionIsMounted = !0, E.updateFeatures(), E.scheduleRenderMicrotask(), A.current && E.animationState && E.animationState.animateChanges());
	}), l(() => {
		E && (!A.current && E.animationState && E.animationState.animateChanges(), A.current &&= (queueMicrotask(() => {
			window.MotionHandoffMarkAsComplete?.(k);
		}), !1), E.enteringChildren = void 0);
	}), E;
}
function p(e, t, n, r) {
	let { layoutId: i, layout: a, drag: o, dragConstraints: c, layoutScroll: l, layoutRoot: u, layoutAnchor: d, layoutCrossfade: f } = t;
	e.projection = new n(e.latestValues, t["data-framer-portal-id"] ? void 0 : m(e.parent)), e.projection.setOptions({
		layoutId: i,
		layout: a,
		alwaysMeasureLayout: !!o || c && s(c),
		visualElement: e,
		animationType: typeof a == "string" ? a : "both",
		initialPromotionConfig: r,
		crossfade: f,
		layoutScroll: l,
		layoutRoot: u,
		layoutAnchor: d
	});
}
function m(e) {
	if (e) return e.options.allowProjection === !1 ? m(e.parent) : e.projection;
}
//#endregion
export { f as useVisualElement };
