// motion-dom/dist/es/frameloop/batcher.js
import { MotionGlobalConfig as e } from "../../../../motion-utils/dist/es/global-config.js";
import { stepsOrder as t } from "./order.js";
import { createRenderStep as n } from "./render-step.js";
//#region node_modules/motion-dom/dist/es/frameloop/batcher.mjs
var r = 40;
function i(i, a) {
	let o = !1, s = !0, c = {
		delta: 0,
		timestamp: 0,
		isProcessing: !1
	}, l = () => o = !0, u = t.reduce((e, t) => (e[t] = n(l, a ? t : void 0), e), {}), { setup: d, read: f, resolveKeyframes: p, preUpdate: m, update: h, preRender: g, render: _, postRender: v } = u, y = () => {
		let t = e.useManualTiming, n = t ? c.timestamp : performance.now();
		o = !1, t || (c.delta = s ? 1e3 / 60 : Math.max(Math.min(n - c.timestamp, r), 1)), c.timestamp = n, c.isProcessing = !0, d.process(c), f.process(c), p.process(c), m.process(c), h.process(c), g.process(c), _.process(c), v.process(c), c.isProcessing = !1, o && a && (s = !1, i(y));
	}, b = () => {
		o = !0, s = !0, c.isProcessing || i(y);
	};
	return {
		schedule: t.reduce((e, t) => {
			let n = u[t];
			return e[t] = (e, t = !1, r = !1) => (o || b(), n.schedule(e, t, r)), e;
		}, {}),
		cancel: (e) => {
			for (let n = 0; n < t.length; n++) u[t[n]].cancel(e);
		},
		state: c,
		steps: u
	};
}
//#endregion
export { i as createRenderBatcher };
