// framer-motion/dist/es/motion/features/animation/exit.js
import { resolveVariant as e } from "../../../../../../motion-dom/dist/es/render/utils/resolve-dynamic-variants.js";
import { Feature as t } from "../../../../../../motion-dom/dist/es/render/Feature.js";
//#region node_modules/framer-motion/dist/es/motion/features/animation/exit.mjs
var n = 0, r = class extends t {
	constructor() {
		super(...arguments), this.id = n++, this.isExitComplete = !1;
	}
	update() {
		if (!this.node.presenceContext) return;
		let { isPresent: t, onExitComplete: n } = this.node.presenceContext, { isPresent: r } = this.node.prevPresenceContext || {};
		if (!this.node.animationState || t === r) return;
		if (t && r === !1) {
			if (this.isExitComplete) {
				let { initial: t, custom: n } = this.node.getProps();
				if (typeof t == "string" || typeof t == "object" && t && !Array.isArray(t)) {
					let r = e(this.node, t, n);
					if (r) {
						let { transition: e, transitionEnd: t, ...n } = r;
						for (let e in n) this.node.getValue(e)?.jump(n[e]);
					}
				}
				this.node.animationState.reset(), this.node.animationState.animateChanges();
			} else this.node.animationState.setActive("exit", !1);
			this.isExitComplete = !1;
			return;
		}
		let i = this.node.animationState.setActive("exit", !t);
		n && !t && i.then(() => {
			this.isExitComplete = !0, n(this.id);
		});
	}
	mount() {
		let { register: e, onExitComplete: t } = this.node.presenceContext || {};
		t && t(this.id), e && (this.unmount = e(this.id));
	}
	unmount() {}
};
//#endregion
export { r as ExitAnimationFeature };
